// @aws-sdk/client-cloudwatch/dist-cjs/protocols/Aws_query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.de_DisableInsightRulesCommand = exports.de_DisableAlarmActionsCommand = exports.de_DescribeInsightRulesCommand = exports.de_DescribeAnomalyDetectorsCommand = exports.de_DescribeAlarmsForMetricCommand = exports.de_DescribeAlarmsCommand = exports.de_DescribeAlarmHistoryCommand = exports.de_DeleteMetricStreamCommand = exports.de_DeleteInsightRulesCommand = exports.de_DeleteDashboardsCommand = exports.de_DeleteAnomalyDetectorCommand = exports.de_DeleteAlarmsCommand = exports.se_UntagResourceCommand = exports.se_TagResourceCommand = exports.se_StopMetricStreamsCommand = exports.se_StartMetricStreamsCommand = exports.se_SetAlarmStateCommand = exports.se_PutMetricStreamCommand = exports.se_PutMetricDataCommand = exports.se_PutMetricAlarmCommand = exports.se_PutManagedInsightRulesCommand = exports.se_PutInsightRuleCommand = exports.se_PutDashboardCommand = exports.se_PutCompositeAlarmCommand = exports.se_PutAnomalyDetectorCommand = exports.se_ListTagsForResourceCommand = exports.se_ListMetricStreamsCommand = exports.se_ListMetricsCommand = exports.se_ListManagedInsightRulesCommand = exports.se_ListDashboardsCommand = exports.se_GetMetricWidgetImageCommand = exports.se_GetMetricStreamCommand = exports.se_GetMetricStatisticsCommand = exports.se_GetMetricDataCommand = exports.se_GetInsightRuleReportCommand = exports.se_GetDashboardCommand = exports.se_EnableInsightRulesCommand = exports.se_EnableAlarmActionsCommand = exports.se_DisableInsightRulesCommand = exports.se_DisableAlarmActionsCommand = exports.se_DescribeInsightRulesCommand = exports.se_DescribeAnomalyDetectorsCommand = exports.se_DescribeAlarmsForMetricCommand = exports.se_DescribeAlarmsCommand = exports.se_DescribeAlarmHistoryCommand = exports.se_DeleteMetricStreamCommand = exports.se_DeleteInsightRulesCommand = exports.se_DeleteDashboardsCommand = exports.se_DeleteAnomalyDetectorCommand = exports.se_DeleteAlarmsCommand = void 0;
exports.de_UntagResourceCommand = exports.de_TagResourceCommand = exports.de_StopMetricStreamsCommand = exports.de_StartMetricStreamsCommand = exports.de_SetAlarmStateCommand = exports.de_PutMetricStreamCommand = exports.de_PutMetricDataCommand = exports.de_PutMetricAlarmCommand = exports.de_PutManagedInsightRulesCommand = exports.de_PutInsightRuleCommand = exports.de_PutDashboardCommand = exports.de_PutCompositeAlarmCommand = exports.de_PutAnomalyDetectorCommand = exports.de_ListTagsForResourceCommand = exports.de_ListMetricStreamsCommand = exports.de_ListMetricsCommand = exports.de_ListManagedInsightRulesCommand = exports.de_ListDashboardsCommand = exports.de_GetMetricWidgetImageCommand = exports.de_GetMetricStreamCommand = exports.de_GetMetricStatisticsCommand = exports.de_GetMetricDataCommand = exports.de_GetInsightRuleReportCommand = exports.de_GetDashboardCommand = exports.de_EnableInsightRulesCommand = exports.de_EnableAlarmActionsCommand = void 0;
const protocol_http_1 = require("@smithy/protocol-http");
const smithy_client_1 = require("@smithy/smithy-client");
const fast_xml_parser_1 = require("fast-xml-parser");
const CloudWatchServiceException_1 = require("../models/CloudWatchServiceException");
const models_0_1 = require("../models/models_0");
const se_DeleteAlarmsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DeleteAlarmsInput(input, context),
        Action: "DeleteAlarms",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DeleteAlarmsCommand = se_DeleteAlarmsCommand;
const se_DeleteAnomalyDetectorCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DeleteAnomalyDetectorInput(input, context),
        Action: "DeleteAnomalyDetector",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DeleteAnomalyDetectorCommand = se_DeleteAnomalyDetectorCommand;
const se_DeleteDashboardsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DeleteDashboardsInput(input, context),
        Action: "DeleteDashboards",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DeleteDashboardsCommand = se_DeleteDashboardsCommand;
const se_DeleteInsightRulesCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DeleteInsightRulesInput(input, context),
        Action: "DeleteInsightRules",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DeleteInsightRulesCommand = se_DeleteInsightRulesCommand;
const se_DeleteMetricStreamCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DeleteMetricStreamInput(input, context),
        Action: "DeleteMetricStream",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DeleteMetricStreamCommand = se_DeleteMetricStreamCommand;
const se_DescribeAlarmHistoryCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DescribeAlarmHistoryInput(input, context),
        Action: "DescribeAlarmHistory",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DescribeAlarmHistoryCommand = se_DescribeAlarmHistoryCommand;
const se_DescribeAlarmsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DescribeAlarmsInput(input, context),
        Action: "DescribeAlarms",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DescribeAlarmsCommand = se_DescribeAlarmsCommand;
const se_DescribeAlarmsForMetricCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DescribeAlarmsForMetricInput(input, context),
        Action: "DescribeAlarmsForMetric",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DescribeAlarmsForMetricCommand = se_DescribeAlarmsForMetricCommand;
const se_DescribeAnomalyDetectorsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DescribeAnomalyDetectorsInput(input, context),
        Action: "DescribeAnomalyDetectors",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DescribeAnomalyDetectorsCommand = se_DescribeAnomalyDetectorsCommand;
const se_DescribeInsightRulesCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DescribeInsightRulesInput(input, context),
        Action: "DescribeInsightRules",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DescribeInsightRulesCommand = se_DescribeInsightRulesCommand;
const se_DisableAlarmActionsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DisableAlarmActionsInput(input, context),
        Action: "DisableAlarmActions",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DisableAlarmActionsCommand = se_DisableAlarmActionsCommand;
const se_DisableInsightRulesCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_DisableInsightRulesInput(input, context),
        Action: "DisableInsightRules",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_DisableInsightRulesCommand = se_DisableInsightRulesCommand;
const se_EnableAlarmActionsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_EnableAlarmActionsInput(input, context),
        Action: "EnableAlarmActions",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_EnableAlarmActionsCommand = se_EnableAlarmActionsCommand;
const se_EnableInsightRulesCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_EnableInsightRulesInput(input, context),
        Action: "EnableInsightRules",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_EnableInsightRulesCommand = se_EnableInsightRulesCommand;
const se_GetDashboardCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_GetDashboardInput(input, context),
        Action: "GetDashboard",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_GetDashboardCommand = se_GetDashboardCommand;
const se_GetInsightRuleReportCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_GetInsightRuleReportInput(input, context),
        Action: "GetInsightRuleReport",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_GetInsightRuleReportCommand = se_GetInsightRuleReportCommand;
const se_GetMetricDataCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_GetMetricDataInput(input, context),
        Action: "GetMetricData",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_GetMetricDataCommand = se_GetMetricDataCommand;
const se_GetMetricStatisticsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_GetMetricStatisticsInput(input, context),
        Action: "GetMetricStatistics",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_GetMetricStatisticsCommand = se_GetMetricStatisticsCommand;
const se_GetMetricStreamCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_GetMetricStreamInput(input, context),
        Action: "GetMetricStream",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_GetMetricStreamCommand = se_GetMetricStreamCommand;
const se_GetMetricWidgetImageCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_GetMetricWidgetImageInput(input, context),
        Action: "GetMetricWidgetImage",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_GetMetricWidgetImageCommand = se_GetMetricWidgetImageCommand;
const se_ListDashboardsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_ListDashboardsInput(input, context),
        Action: "ListDashboards",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_ListDashboardsCommand = se_ListDashboardsCommand;
const se_ListManagedInsightRulesCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_ListManagedInsightRulesInput(input, context),
        Action: "ListManagedInsightRules",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_ListManagedInsightRulesCommand = se_ListManagedInsightRulesCommand;
const se_ListMetricsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_ListMetricsInput(input, context),
        Action: "ListMetrics",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_ListMetricsCommand = se_ListMetricsCommand;
const se_ListMetricStreamsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_ListMetricStreamsInput(input, context),
        Action: "ListMetricStreams",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_ListMetricStreamsCommand = se_ListMetricStreamsCommand;
const se_ListTagsForResourceCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_ListTagsForResourceInput(input, context),
        Action: "ListTagsForResource",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_ListTagsForResourceCommand = se_ListTagsForResourceCommand;
const se_PutAnomalyDetectorCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutAnomalyDetectorInput(input, context),
        Action: "PutAnomalyDetector",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutAnomalyDetectorCommand = se_PutAnomalyDetectorCommand;
const se_PutCompositeAlarmCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutCompositeAlarmInput(input, context),
        Action: "PutCompositeAlarm",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutCompositeAlarmCommand = se_PutCompositeAlarmCommand;
const se_PutDashboardCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutDashboardInput(input, context),
        Action: "PutDashboard",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutDashboardCommand = se_PutDashboardCommand;
const se_PutInsightRuleCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutInsightRuleInput(input, context),
        Action: "PutInsightRule",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutInsightRuleCommand = se_PutInsightRuleCommand;
const se_PutManagedInsightRulesCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutManagedInsightRulesInput(input, context),
        Action: "PutManagedInsightRules",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutManagedInsightRulesCommand = se_PutManagedInsightRulesCommand;
const se_PutMetricAlarmCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutMetricAlarmInput(input, context),
        Action: "PutMetricAlarm",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutMetricAlarmCommand = se_PutMetricAlarmCommand;
const se_PutMetricDataCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutMetricDataInput(input, context),
        Action: "PutMetricData",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutMetricDataCommand = se_PutMetricDataCommand;
const se_PutMetricStreamCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_PutMetricStreamInput(input, context),
        Action: "PutMetricStream",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_PutMetricStreamCommand = se_PutMetricStreamCommand;
const se_SetAlarmStateCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_SetAlarmStateInput(input, context),
        Action: "SetAlarmState",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_SetAlarmStateCommand = se_SetAlarmStateCommand;
const se_StartMetricStreamsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_StartMetricStreamsInput(input, context),
        Action: "StartMetricStreams",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_StartMetricStreamsCommand = se_StartMetricStreamsCommand;
const se_StopMetricStreamsCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_StopMetricStreamsInput(input, context),
        Action: "StopMetricStreams",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_StopMetricStreamsCommand = se_StopMetricStreamsCommand;
const se_TagResourceCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_TagResourceInput(input, context),
        Action: "TagResource",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_TagResourceCommand = se_TagResourceCommand;
const se_UntagResourceCommand = async (input, context) => {
    const headers = SHARED_HEADERS;
    let body;
    body = buildFormUrlencodedString({
        ...se_UntagResourceInput(input, context),
        Action: "UntagResource",
        Version: "2010-08-01",
    });
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
exports.se_UntagResourceCommand = se_UntagResourceCommand;
const de_DeleteAlarmsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DeleteAlarmsCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_DeleteAlarmsCommand = de_DeleteAlarmsCommand;
const de_DeleteAlarmsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "ResourceNotFound":
        case "com.amazonaws.cloudwatch#ResourceNotFound":
            throw await de_ResourceNotFoundRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DeleteAnomalyDetectorCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DeleteAnomalyDetectorCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DeleteAnomalyDetectorOutput(data.DeleteAnomalyDetectorResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DeleteAnomalyDetectorCommand = de_DeleteAnomalyDetectorCommand;
const de_DeleteAnomalyDetectorCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        case "ResourceNotFoundException":
        case "com.amazonaws.cloudwatch#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DeleteDashboardsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DeleteDashboardsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DeleteDashboardsOutput(data.DeleteDashboardsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DeleteDashboardsCommand = de_DeleteDashboardsCommand;
const de_DeleteDashboardsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "ResourceNotFound":
        case "com.amazonaws.cloudwatch#DashboardNotFoundError":
            throw await de_DashboardNotFoundErrorRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DeleteInsightRulesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DeleteInsightRulesCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DeleteInsightRulesOutput(data.DeleteInsightRulesResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DeleteInsightRulesCommand = de_DeleteInsightRulesCommand;
const de_DeleteInsightRulesCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DeleteMetricStreamCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DeleteMetricStreamCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DeleteMetricStreamOutput(data.DeleteMetricStreamResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DeleteMetricStreamCommand = de_DeleteMetricStreamCommand;
const de_DeleteMetricStreamCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DescribeAlarmHistoryCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DescribeAlarmHistoryCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DescribeAlarmHistoryOutput(data.DescribeAlarmHistoryResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DescribeAlarmHistoryCommand = de_DescribeAlarmHistoryCommand;
const de_DescribeAlarmHistoryCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DescribeAlarmsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DescribeAlarmsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DescribeAlarmsOutput(data.DescribeAlarmsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DescribeAlarmsCommand = de_DescribeAlarmsCommand;
const de_DescribeAlarmsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DescribeAlarmsForMetricCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DescribeAlarmsForMetricCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DescribeAlarmsForMetricOutput(data.DescribeAlarmsForMetricResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DescribeAlarmsForMetricCommand = de_DescribeAlarmsForMetricCommand;
const de_DescribeAlarmsForMetricCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    const parsedBody = parsedOutput.body;
    return throwDefaultError({
        output,
        parsedBody: parsedBody.Error,
        errorCode,
    });
};
const de_DescribeAnomalyDetectorsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DescribeAnomalyDetectorsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DescribeAnomalyDetectorsOutput(data.DescribeAnomalyDetectorsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DescribeAnomalyDetectorsCommand = de_DescribeAnomalyDetectorsCommand;
const de_DescribeAnomalyDetectorsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DescribeInsightRulesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DescribeInsightRulesCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DescribeInsightRulesOutput(data.DescribeInsightRulesResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DescribeInsightRulesCommand = de_DescribeInsightRulesCommand;
const de_DescribeInsightRulesCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_DisableAlarmActionsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DisableAlarmActionsCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_DisableAlarmActionsCommand = de_DisableAlarmActionsCommand;
const de_DisableAlarmActionsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    const parsedBody = parsedOutput.body;
    return throwDefaultError({
        output,
        parsedBody: parsedBody.Error,
        errorCode,
    });
};
const de_DisableInsightRulesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_DisableInsightRulesCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_DisableInsightRulesOutput(data.DisableInsightRulesResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_DisableInsightRulesCommand = de_DisableInsightRulesCommand;
const de_DisableInsightRulesCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_EnableAlarmActionsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_EnableAlarmActionsCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_EnableAlarmActionsCommand = de_EnableAlarmActionsCommand;
const de_EnableAlarmActionsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    const parsedBody = parsedOutput.body;
    return throwDefaultError({
        output,
        parsedBody: parsedBody.Error,
        errorCode,
    });
};
const de_EnableInsightRulesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_EnableInsightRulesCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_EnableInsightRulesOutput(data.EnableInsightRulesResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_EnableInsightRulesCommand = de_EnableInsightRulesCommand;
const de_EnableInsightRulesCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "LimitExceededException":
        case "com.amazonaws.cloudwatch#LimitExceededException":
            throw await de_LimitExceededExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_GetDashboardCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_GetDashboardCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_GetDashboardOutput(data.GetDashboardResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_GetDashboardCommand = de_GetDashboardCommand;
const de_GetDashboardCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "ResourceNotFound":
        case "com.amazonaws.cloudwatch#DashboardNotFoundError":
            throw await de_DashboardNotFoundErrorRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_GetInsightRuleReportCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_GetInsightRuleReportCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_GetInsightRuleReportOutput(data.GetInsightRuleReportResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_GetInsightRuleReportCommand = de_GetInsightRuleReportCommand;
const de_GetInsightRuleReportCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        case "ResourceNotFoundException":
        case "com.amazonaws.cloudwatch#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_GetMetricDataCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_GetMetricDataCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_GetMetricDataOutput(data.GetMetricDataResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_GetMetricDataCommand = de_GetMetricDataCommand;
const de_GetMetricDataCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_GetMetricStatisticsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_GetMetricStatisticsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_GetMetricStatisticsOutput(data.GetMetricStatisticsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_GetMetricStatisticsCommand = de_GetMetricStatisticsCommand;
const de_GetMetricStatisticsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_GetMetricStreamCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_GetMetricStreamCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_GetMetricStreamOutput(data.GetMetricStreamResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_GetMetricStreamCommand = de_GetMetricStreamCommand;
const de_GetMetricStreamCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        case "ResourceNotFoundException":
        case "com.amazonaws.cloudwatch#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_GetMetricWidgetImageCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_GetMetricWidgetImageCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_GetMetricWidgetImageOutput(data.GetMetricWidgetImageResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_GetMetricWidgetImageCommand = de_GetMetricWidgetImageCommand;
const de_GetMetricWidgetImageCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    const parsedBody = parsedOutput.body;
    return throwDefaultError({
        output,
        parsedBody: parsedBody.Error,
        errorCode,
    });
};
const de_ListDashboardsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_ListDashboardsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_ListDashboardsOutput(data.ListDashboardsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_ListDashboardsCommand = de_ListDashboardsCommand;
const de_ListDashboardsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_ListManagedInsightRulesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_ListManagedInsightRulesCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_ListManagedInsightRulesOutput(data.ListManagedInsightRulesResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_ListManagedInsightRulesCommand = de_ListManagedInsightRulesCommand;
const de_ListManagedInsightRulesCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_ListMetricsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_ListMetricsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_ListMetricsOutput(data.ListMetricsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_ListMetricsCommand = de_ListMetricsCommand;
const de_ListMetricsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_ListMetricStreamsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_ListMetricStreamsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_ListMetricStreamsOutput(data.ListMetricStreamsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_ListMetricStreamsCommand = de_ListMetricStreamsCommand;
const de_ListMetricStreamsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidNextToken":
        case "com.amazonaws.cloudwatch#InvalidNextToken":
            throw await de_InvalidNextTokenRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_ListTagsForResourceCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_ListTagsForResourceCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_ListTagsForResourceOutput(data.ListTagsForResourceResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_ListTagsForResourceCommand = de_ListTagsForResourceCommand;
const de_ListTagsForResourceCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "ResourceNotFoundException":
        case "com.amazonaws.cloudwatch#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutAnomalyDetectorCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutAnomalyDetectorCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_PutAnomalyDetectorOutput(data.PutAnomalyDetectorResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_PutAnomalyDetectorCommand = de_PutAnomalyDetectorCommand;
const de_PutAnomalyDetectorCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "LimitExceededException":
        case "com.amazonaws.cloudwatch#LimitExceededException":
            throw await de_LimitExceededExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutCompositeAlarmCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutCompositeAlarmCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_PutCompositeAlarmCommand = de_PutCompositeAlarmCommand;
const de_PutCompositeAlarmCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "LimitExceeded":
        case "com.amazonaws.cloudwatch#LimitExceededFault":
            throw await de_LimitExceededFaultRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutDashboardCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutDashboardCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_PutDashboardOutput(data.PutDashboardResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_PutDashboardCommand = de_PutDashboardCommand;
const de_PutDashboardCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterInput":
        case "com.amazonaws.cloudwatch#DashboardInvalidInputError":
            throw await de_DashboardInvalidInputErrorRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutInsightRuleCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutInsightRuleCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_PutInsightRuleOutput(data.PutInsightRuleResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_PutInsightRuleCommand = de_PutInsightRuleCommand;
const de_PutInsightRuleCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "LimitExceededException":
        case "com.amazonaws.cloudwatch#LimitExceededException":
            throw await de_LimitExceededExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutManagedInsightRulesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutManagedInsightRulesCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_PutManagedInsightRulesOutput(data.PutManagedInsightRulesResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_PutManagedInsightRulesCommand = de_PutManagedInsightRulesCommand;
const de_PutManagedInsightRulesCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutMetricAlarmCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutMetricAlarmCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_PutMetricAlarmCommand = de_PutMetricAlarmCommand;
const de_PutMetricAlarmCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "LimitExceeded":
        case "com.amazonaws.cloudwatch#LimitExceededFault":
            throw await de_LimitExceededFaultRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutMetricDataCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutMetricDataCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_PutMetricDataCommand = de_PutMetricDataCommand;
const de_PutMetricDataCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_PutMetricStreamCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_PutMetricStreamCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_PutMetricStreamOutput(data.PutMetricStreamResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_PutMetricStreamCommand = de_PutMetricStreamCommand;
const de_PutMetricStreamCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "ConcurrentModificationException":
        case "com.amazonaws.cloudwatch#ConcurrentModificationException":
            throw await de_ConcurrentModificationExceptionRes(parsedOutput, context);
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterCombination":
        case "com.amazonaws.cloudwatch#InvalidParameterCombinationException":
            throw await de_InvalidParameterCombinationExceptionRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_SetAlarmStateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_SetAlarmStateCommandError(output, context);
    }
    await (0, smithy_client_1.collectBody)(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
exports.de_SetAlarmStateCommand = de_SetAlarmStateCommand;
const de_SetAlarmStateCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidFormat":
        case "com.amazonaws.cloudwatch#InvalidFormatFault":
            throw await de_InvalidFormatFaultRes(parsedOutput, context);
        case "ResourceNotFound":
        case "com.amazonaws.cloudwatch#ResourceNotFound":
            throw await de_ResourceNotFoundRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_StartMetricStreamsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_StartMetricStreamsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_StartMetricStreamsOutput(data.StartMetricStreamsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_StartMetricStreamsCommand = de_StartMetricStreamsCommand;
const de_StartMetricStreamsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_StopMetricStreamsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_StopMetricStreamsCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_StopMetricStreamsOutput(data.StopMetricStreamsResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_StopMetricStreamsCommand = de_StopMetricStreamsCommand;
const de_StopMetricStreamsCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "MissingParameter":
        case "com.amazonaws.cloudwatch#MissingRequiredParameterException":
            throw await de_MissingRequiredParameterExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_TagResourceCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_TagResourceCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_TagResourceOutput(data.TagResourceResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_TagResourceCommand = de_TagResourceCommand;
const de_TagResourceCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "ConcurrentModificationException":
        case "com.amazonaws.cloudwatch#ConcurrentModificationException":
            throw await de_ConcurrentModificationExceptionRes(parsedOutput, context);
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "ResourceNotFoundException":
        case "com.amazonaws.cloudwatch#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_UntagResourceCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_UntagResourceCommandError(output, context);
    }
    const data = await parseBody(output.body, context);
    let contents = {};
    contents = de_UntagResourceOutput(data.UntagResourceResult, context);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
exports.de_UntagResourceCommand = de_UntagResourceCommand;
const de_UntagResourceCommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await parseErrorBody(output.body, context),
    };
    const errorCode = loadQueryErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "ConcurrentModificationException":
        case "com.amazonaws.cloudwatch#ConcurrentModificationException":
            throw await de_ConcurrentModificationExceptionRes(parsedOutput, context);
        case "InternalServiceError":
        case "com.amazonaws.cloudwatch#InternalServiceFault":
            throw await de_InternalServiceFaultRes(parsedOutput, context);
        case "InvalidParameterValue":
        case "com.amazonaws.cloudwatch#InvalidParameterValueException":
            throw await de_InvalidParameterValueExceptionRes(parsedOutput, context);
        case "ResourceNotFoundException":
        case "com.amazonaws.cloudwatch#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput, context);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody: parsedBody.Error,
                errorCode,
            });
    }
};
const de_ConcurrentModificationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_ConcurrentModificationException(body.Error, context);
    const exception = new models_0_1.ConcurrentModificationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_DashboardInvalidInputErrorRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_DashboardInvalidInputError(body.Error, context);
    const exception = new models_0_1.DashboardInvalidInputError({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_DashboardNotFoundErrorRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_DashboardNotFoundError(body.Error, context);
    const exception = new models_0_1.DashboardNotFoundError({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_InternalServiceFaultRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_InternalServiceFault(body.Error, context);
    const exception = new models_0_1.InternalServiceFault({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_InvalidFormatFaultRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_InvalidFormatFault(body.Error, context);
    const exception = new models_0_1.InvalidFormatFault({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_InvalidNextTokenRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_InvalidNextToken(body.Error, context);
    const exception = new models_0_1.InvalidNextToken({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_InvalidParameterCombinationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_InvalidParameterCombinationException(body.Error, context);
    const exception = new models_0_1.InvalidParameterCombinationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_InvalidParameterValueExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_InvalidParameterValueException(body.Error, context);
    const exception = new models_0_1.InvalidParameterValueException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_LimitExceededExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_LimitExceededException(body.Error, context);
    const exception = new models_0_1.LimitExceededException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_LimitExceededFaultRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_LimitExceededFault(body.Error, context);
    const exception = new models_0_1.LimitExceededFault({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_MissingRequiredParameterExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_MissingRequiredParameterException(body.Error, context);
    const exception = new models_0_1.MissingRequiredParameterException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_ResourceNotFoundRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_ResourceNotFound(body.Error, context);
    const exception = new models_0_1.ResourceNotFound({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const de_ResourceNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = de_ResourceNotFoundException(body.Error, context);
    const exception = new models_0_1.ResourceNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return (0, smithy_client_1.decorateServiceException)(exception, body);
};
const se_AlarmNames = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_AlarmTypes = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_AnomalyDetectorConfiguration = (input, context) => {
    const entries = {};
    if (input.ExcludedTimeRanges != null) {
        const memberEntries = se_AnomalyDetectorExcludedTimeRanges(input.ExcludedTimeRanges, context);
        if (input.ExcludedTimeRanges?.length === 0) {
            entries.ExcludedTimeRanges = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `ExcludedTimeRanges.${key}`;
            entries[loc] = value;
        });
    }
    if (input.MetricTimezone != null) {
        entries["MetricTimezone"] = input.MetricTimezone;
    }
    return entries;
};
const se_AnomalyDetectorExcludedTimeRanges = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_Range(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_AnomalyDetectorTypes = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_Counts = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = (0, smithy_client_1.serializeFloat)(entry);
        counter++;
    }
    return entries;
};
const se_DashboardNames = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_DeleteAlarmsInput = (input, context) => {
    const entries = {};
    if (input.AlarmNames != null) {
        const memberEntries = se_AlarmNames(input.AlarmNames, context);
        if (input.AlarmNames?.length === 0) {
            entries.AlarmNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_DeleteAnomalyDetectorInput = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Stat != null) {
        entries["Stat"] = input.Stat;
    }
    if (input.SingleMetricAnomalyDetector != null) {
        const memberEntries = se_SingleMetricAnomalyDetector(input.SingleMetricAnomalyDetector, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `SingleMetricAnomalyDetector.${key}`;
            entries[loc] = value;
        });
    }
    if (input.MetricMathAnomalyDetector != null) {
        const memberEntries = se_MetricMathAnomalyDetector(input.MetricMathAnomalyDetector, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricMathAnomalyDetector.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_DeleteDashboardsInput = (input, context) => {
    const entries = {};
    if (input.DashboardNames != null) {
        const memberEntries = se_DashboardNames(input.DashboardNames, context);
        if (input.DashboardNames?.length === 0) {
            entries.DashboardNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `DashboardNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_DeleteInsightRulesInput = (input, context) => {
    const entries = {};
    if (input.RuleNames != null) {
        const memberEntries = se_InsightRuleNames(input.RuleNames, context);
        if (input.RuleNames?.length === 0) {
            entries.RuleNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `RuleNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_DeleteMetricStreamInput = (input, context) => {
    const entries = {};
    if (input.Name != null) {
        entries["Name"] = input.Name;
    }
    return entries;
};
const se_DescribeAlarmHistoryInput = (input, context) => {
    const entries = {};
    if (input.AlarmName != null) {
        entries["AlarmName"] = input.AlarmName;
    }
    if (input.AlarmTypes != null) {
        const memberEntries = se_AlarmTypes(input.AlarmTypes, context);
        if (input.AlarmTypes?.length === 0) {
            entries.AlarmTypes = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmTypes.${key}`;
            entries[loc] = value;
        });
    }
    if (input.HistoryItemType != null) {
        entries["HistoryItemType"] = input.HistoryItemType;
    }
    if (input.StartDate != null) {
        entries["StartDate"] = input.StartDate.toISOString().split(".")[0] + "Z";
    }
    if (input.EndDate != null) {
        entries["EndDate"] = input.EndDate.toISOString().split(".")[0] + "Z";
    }
    if (input.MaxRecords != null) {
        entries["MaxRecords"] = input.MaxRecords;
    }
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.ScanBy != null) {
        entries["ScanBy"] = input.ScanBy;
    }
    return entries;
};
const se_DescribeAlarmsForMetricInput = (input, context) => {
    const entries = {};
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.Statistic != null) {
        entries["Statistic"] = input.Statistic;
    }
    if (input.ExtendedStatistic != null) {
        entries["ExtendedStatistic"] = input.ExtendedStatistic;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Period != null) {
        entries["Period"] = input.Period;
    }
    if (input.Unit != null) {
        entries["Unit"] = input.Unit;
    }
    return entries;
};
const se_DescribeAlarmsInput = (input, context) => {
    const entries = {};
    if (input.AlarmNames != null) {
        const memberEntries = se_AlarmNames(input.AlarmNames, context);
        if (input.AlarmNames?.length === 0) {
            entries.AlarmNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmNames.${key}`;
            entries[loc] = value;
        });
    }
    if (input.AlarmNamePrefix != null) {
        entries["AlarmNamePrefix"] = input.AlarmNamePrefix;
    }
    if (input.AlarmTypes != null) {
        const memberEntries = se_AlarmTypes(input.AlarmTypes, context);
        if (input.AlarmTypes?.length === 0) {
            entries.AlarmTypes = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmTypes.${key}`;
            entries[loc] = value;
        });
    }
    if (input.ChildrenOfAlarmName != null) {
        entries["ChildrenOfAlarmName"] = input.ChildrenOfAlarmName;
    }
    if (input.ParentsOfAlarmName != null) {
        entries["ParentsOfAlarmName"] = input.ParentsOfAlarmName;
    }
    if (input.StateValue != null) {
        entries["StateValue"] = input.StateValue;
    }
    if (input.ActionPrefix != null) {
        entries["ActionPrefix"] = input.ActionPrefix;
    }
    if (input.MaxRecords != null) {
        entries["MaxRecords"] = input.MaxRecords;
    }
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    return entries;
};
const se_DescribeAnomalyDetectorsInput = (input, context) => {
    const entries = {};
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.MaxResults != null) {
        entries["MaxResults"] = input.MaxResults;
    }
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.AnomalyDetectorTypes != null) {
        const memberEntries = se_AnomalyDetectorTypes(input.AnomalyDetectorTypes, context);
        if (input.AnomalyDetectorTypes?.length === 0) {
            entries.AnomalyDetectorTypes = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AnomalyDetectorTypes.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_DescribeInsightRulesInput = (input, context) => {
    const entries = {};
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.MaxResults != null) {
        entries["MaxResults"] = input.MaxResults;
    }
    return entries;
};
const se_Dimension = (input, context) => {
    const entries = {};
    if (input.Name != null) {
        entries["Name"] = input.Name;
    }
    if (input.Value != null) {
        entries["Value"] = input.Value;
    }
    return entries;
};
const se_DimensionFilter = (input, context) => {
    const entries = {};
    if (input.Name != null) {
        entries["Name"] = input.Name;
    }
    if (input.Value != null) {
        entries["Value"] = input.Value;
    }
    return entries;
};
const se_DimensionFilters = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_DimensionFilter(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_Dimensions = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_Dimension(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_DisableAlarmActionsInput = (input, context) => {
    const entries = {};
    if (input.AlarmNames != null) {
        const memberEntries = se_AlarmNames(input.AlarmNames, context);
        if (input.AlarmNames?.length === 0) {
            entries.AlarmNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_DisableInsightRulesInput = (input, context) => {
    const entries = {};
    if (input.RuleNames != null) {
        const memberEntries = se_InsightRuleNames(input.RuleNames, context);
        if (input.RuleNames?.length === 0) {
            entries.RuleNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `RuleNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_EnableAlarmActionsInput = (input, context) => {
    const entries = {};
    if (input.AlarmNames != null) {
        const memberEntries = se_AlarmNames(input.AlarmNames, context);
        if (input.AlarmNames?.length === 0) {
            entries.AlarmNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_EnableInsightRulesInput = (input, context) => {
    const entries = {};
    if (input.RuleNames != null) {
        const memberEntries = se_InsightRuleNames(input.RuleNames, context);
        if (input.RuleNames?.length === 0) {
            entries.RuleNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `RuleNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_ExtendedStatistics = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_GetDashboardInput = (input, context) => {
    const entries = {};
    if (input.DashboardName != null) {
        entries["DashboardName"] = input.DashboardName;
    }
    return entries;
};
const se_GetInsightRuleReportInput = (input, context) => {
    const entries = {};
    if (input.RuleName != null) {
        entries["RuleName"] = input.RuleName;
    }
    if (input.StartTime != null) {
        entries["StartTime"] = input.StartTime.toISOString().split(".")[0] + "Z";
    }
    if (input.EndTime != null) {
        entries["EndTime"] = input.EndTime.toISOString().split(".")[0] + "Z";
    }
    if (input.Period != null) {
        entries["Period"] = input.Period;
    }
    if (input.MaxContributorCount != null) {
        entries["MaxContributorCount"] = input.MaxContributorCount;
    }
    if (input.Metrics != null) {
        const memberEntries = se_InsightRuleMetricList(input.Metrics, context);
        if (input.Metrics?.length === 0) {
            entries.Metrics = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Metrics.${key}`;
            entries[loc] = value;
        });
    }
    if (input.OrderBy != null) {
        entries["OrderBy"] = input.OrderBy;
    }
    return entries;
};
const se_GetMetricDataInput = (input, context) => {
    const entries = {};
    if (input.MetricDataQueries != null) {
        const memberEntries = se_MetricDataQueries(input.MetricDataQueries, context);
        if (input.MetricDataQueries?.length === 0) {
            entries.MetricDataQueries = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricDataQueries.${key}`;
            entries[loc] = value;
        });
    }
    if (input.StartTime != null) {
        entries["StartTime"] = input.StartTime.toISOString().split(".")[0] + "Z";
    }
    if (input.EndTime != null) {
        entries["EndTime"] = input.EndTime.toISOString().split(".")[0] + "Z";
    }
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.ScanBy != null) {
        entries["ScanBy"] = input.ScanBy;
    }
    if (input.MaxDatapoints != null) {
        entries["MaxDatapoints"] = input.MaxDatapoints;
    }
    if (input.LabelOptions != null) {
        const memberEntries = se_LabelOptions(input.LabelOptions, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `LabelOptions.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_GetMetricStatisticsInput = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.StartTime != null) {
        entries["StartTime"] = input.StartTime.toISOString().split(".")[0] + "Z";
    }
    if (input.EndTime != null) {
        entries["EndTime"] = input.EndTime.toISOString().split(".")[0] + "Z";
    }
    if (input.Period != null) {
        entries["Period"] = input.Period;
    }
    if (input.Statistics != null) {
        const memberEntries = se_Statistics(input.Statistics, context);
        if (input.Statistics?.length === 0) {
            entries.Statistics = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Statistics.${key}`;
            entries[loc] = value;
        });
    }
    if (input.ExtendedStatistics != null) {
        const memberEntries = se_ExtendedStatistics(input.ExtendedStatistics, context);
        if (input.ExtendedStatistics?.length === 0) {
            entries.ExtendedStatistics = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `ExtendedStatistics.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Unit != null) {
        entries["Unit"] = input.Unit;
    }
    return entries;
};
const se_GetMetricStreamInput = (input, context) => {
    const entries = {};
    if (input.Name != null) {
        entries["Name"] = input.Name;
    }
    return entries;
};
const se_GetMetricWidgetImageInput = (input, context) => {
    const entries = {};
    if (input.MetricWidget != null) {
        entries["MetricWidget"] = input.MetricWidget;
    }
    if (input.OutputFormat != null) {
        entries["OutputFormat"] = input.OutputFormat;
    }
    return entries;
};
const se_InsightRuleMetricList = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_InsightRuleNames = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_LabelOptions = (input, context) => {
    const entries = {};
    if (input.Timezone != null) {
        entries["Timezone"] = input.Timezone;
    }
    return entries;
};
const se_ListDashboardsInput = (input, context) => {
    const entries = {};
    if (input.DashboardNamePrefix != null) {
        entries["DashboardNamePrefix"] = input.DashboardNamePrefix;
    }
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    return entries;
};
const se_ListManagedInsightRulesInput = (input, context) => {
    const entries = {};
    if (input.ResourceARN != null) {
        entries["ResourceARN"] = input.ResourceARN;
    }
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.MaxResults != null) {
        entries["MaxResults"] = input.MaxResults;
    }
    return entries;
};
const se_ListMetricsInput = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_DimensionFilters(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.RecentlyActive != null) {
        entries["RecentlyActive"] = input.RecentlyActive;
    }
    if (input.IncludeLinkedAccounts != null) {
        entries["IncludeLinkedAccounts"] = input.IncludeLinkedAccounts;
    }
    if (input.OwningAccount != null) {
        entries["OwningAccount"] = input.OwningAccount;
    }
    return entries;
};
const se_ListMetricStreamsInput = (input, context) => {
    const entries = {};
    if (input.NextToken != null) {
        entries["NextToken"] = input.NextToken;
    }
    if (input.MaxResults != null) {
        entries["MaxResults"] = input.MaxResults;
    }
    return entries;
};
const se_ListTagsForResourceInput = (input, context) => {
    const entries = {};
    if (input.ResourceARN != null) {
        entries["ResourceARN"] = input.ResourceARN;
    }
    return entries;
};
const se_ManagedRule = (input, context) => {
    const entries = {};
    if (input.TemplateName != null) {
        entries["TemplateName"] = input.TemplateName;
    }
    if (input.ResourceARN != null) {
        entries["ResourceARN"] = input.ResourceARN;
    }
    if (input.Tags != null) {
        const memberEntries = se_TagList(input.Tags, context);
        if (input.Tags?.length === 0) {
            entries.Tags = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Tags.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_ManagedRules = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_ManagedRule(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_Metric = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_MetricData = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_MetricDatum(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_MetricDataQueries = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_MetricDataQuery(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_MetricDataQuery = (input, context) => {
    const entries = {};
    if (input.Id != null) {
        entries["Id"] = input.Id;
    }
    if (input.MetricStat != null) {
        const memberEntries = se_MetricStat(input.MetricStat, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricStat.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Expression != null) {
        entries["Expression"] = input.Expression;
    }
    if (input.Label != null) {
        entries["Label"] = input.Label;
    }
    if (input.ReturnData != null) {
        entries["ReturnData"] = input.ReturnData;
    }
    if (input.Period != null) {
        entries["Period"] = input.Period;
    }
    if (input.AccountId != null) {
        entries["AccountId"] = input.AccountId;
    }
    return entries;
};
const se_MetricDatum = (input, context) => {
    const entries = {};
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Timestamp != null) {
        entries["Timestamp"] = input.Timestamp.toISOString().split(".")[0] + "Z";
    }
    if (input.Value != null) {
        entries["Value"] = (0, smithy_client_1.serializeFloat)(input.Value);
    }
    if (input.StatisticValues != null) {
        const memberEntries = se_StatisticSet(input.StatisticValues, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `StatisticValues.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Values != null) {
        const memberEntries = se_Values(input.Values, context);
        if (input.Values?.length === 0) {
            entries.Values = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Values.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Counts != null) {
        const memberEntries = se_Counts(input.Counts, context);
        if (input.Counts?.length === 0) {
            entries.Counts = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Counts.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Unit != null) {
        entries["Unit"] = input.Unit;
    }
    if (input.StorageResolution != null) {
        entries["StorageResolution"] = input.StorageResolution;
    }
    return entries;
};
const se_MetricMathAnomalyDetector = (input, context) => {
    const entries = {};
    if (input.MetricDataQueries != null) {
        const memberEntries = se_MetricDataQueries(input.MetricDataQueries, context);
        if (input.MetricDataQueries?.length === 0) {
            entries.MetricDataQueries = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricDataQueries.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_MetricStat = (input, context) => {
    const entries = {};
    if (input.Metric != null) {
        const memberEntries = se_Metric(input.Metric, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Metric.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Period != null) {
        entries["Period"] = input.Period;
    }
    if (input.Stat != null) {
        entries["Stat"] = input.Stat;
    }
    if (input.Unit != null) {
        entries["Unit"] = input.Unit;
    }
    return entries;
};
const se_MetricStreamFilter = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricNames != null) {
        const memberEntries = se_MetricStreamFilterMetricNames(input.MetricNames, context);
        if (input.MetricNames?.length === 0) {
            entries.MetricNames = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricNames.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_MetricStreamFilterMetricNames = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_MetricStreamFilters = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_MetricStreamFilter(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_MetricStreamNames = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_MetricStreamStatisticsAdditionalStatistics = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_MetricStreamStatisticsConfiguration = (input, context) => {
    const entries = {};
    if (input.IncludeMetrics != null) {
        const memberEntries = se_MetricStreamStatisticsIncludeMetrics(input.IncludeMetrics, context);
        if (input.IncludeMetrics?.length === 0) {
            entries.IncludeMetrics = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `IncludeMetrics.${key}`;
            entries[loc] = value;
        });
    }
    if (input.AdditionalStatistics != null) {
        const memberEntries = se_MetricStreamStatisticsAdditionalStatistics(input.AdditionalStatistics, context);
        if (input.AdditionalStatistics?.length === 0) {
            entries.AdditionalStatistics = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AdditionalStatistics.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_MetricStreamStatisticsConfigurations = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_MetricStreamStatisticsConfiguration(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_MetricStreamStatisticsIncludeMetrics = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_MetricStreamStatisticsMetric(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_MetricStreamStatisticsMetric = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    return entries;
};
const se_PutAnomalyDetectorInput = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Stat != null) {
        entries["Stat"] = input.Stat;
    }
    if (input.Configuration != null) {
        const memberEntries = se_AnomalyDetectorConfiguration(input.Configuration, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Configuration.${key}`;
            entries[loc] = value;
        });
    }
    if (input.SingleMetricAnomalyDetector != null) {
        const memberEntries = se_SingleMetricAnomalyDetector(input.SingleMetricAnomalyDetector, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `SingleMetricAnomalyDetector.${key}`;
            entries[loc] = value;
        });
    }
    if (input.MetricMathAnomalyDetector != null) {
        const memberEntries = se_MetricMathAnomalyDetector(input.MetricMathAnomalyDetector, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricMathAnomalyDetector.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_PutCompositeAlarmInput = (input, context) => {
    const entries = {};
    if (input.ActionsEnabled != null) {
        entries["ActionsEnabled"] = input.ActionsEnabled;
    }
    if (input.AlarmActions != null) {
        const memberEntries = se_ResourceList(input.AlarmActions, context);
        if (input.AlarmActions?.length === 0) {
            entries.AlarmActions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmActions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.AlarmDescription != null) {
        entries["AlarmDescription"] = input.AlarmDescription;
    }
    if (input.AlarmName != null) {
        entries["AlarmName"] = input.AlarmName;
    }
    if (input.AlarmRule != null) {
        entries["AlarmRule"] = input.AlarmRule;
    }
    if (input.InsufficientDataActions != null) {
        const memberEntries = se_ResourceList(input.InsufficientDataActions, context);
        if (input.InsufficientDataActions?.length === 0) {
            entries.InsufficientDataActions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `InsufficientDataActions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.OKActions != null) {
        const memberEntries = se_ResourceList(input.OKActions, context);
        if (input.OKActions?.length === 0) {
            entries.OKActions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `OKActions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Tags != null) {
        const memberEntries = se_TagList(input.Tags, context);
        if (input.Tags?.length === 0) {
            entries.Tags = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Tags.${key}`;
            entries[loc] = value;
        });
    }
    if (input.ActionsSuppressor != null) {
        entries["ActionsSuppressor"] = input.ActionsSuppressor;
    }
    if (input.ActionsSuppressorWaitPeriod != null) {
        entries["ActionsSuppressorWaitPeriod"] = input.ActionsSuppressorWaitPeriod;
    }
    if (input.ActionsSuppressorExtensionPeriod != null) {
        entries["ActionsSuppressorExtensionPeriod"] = input.ActionsSuppressorExtensionPeriod;
    }
    return entries;
};
const se_PutDashboardInput = (input, context) => {
    const entries = {};
    if (input.DashboardName != null) {
        entries["DashboardName"] = input.DashboardName;
    }
    if (input.DashboardBody != null) {
        entries["DashboardBody"] = input.DashboardBody;
    }
    return entries;
};
const se_PutInsightRuleInput = (input, context) => {
    const entries = {};
    if (input.RuleName != null) {
        entries["RuleName"] = input.RuleName;
    }
    if (input.RuleState != null) {
        entries["RuleState"] = input.RuleState;
    }
    if (input.RuleDefinition != null) {
        entries["RuleDefinition"] = input.RuleDefinition;
    }
    if (input.Tags != null) {
        const memberEntries = se_TagList(input.Tags, context);
        if (input.Tags?.length === 0) {
            entries.Tags = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Tags.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_PutManagedInsightRulesInput = (input, context) => {
    const entries = {};
    if (input.ManagedRules != null) {
        const memberEntries = se_ManagedRules(input.ManagedRules, context);
        if (input.ManagedRules?.length === 0) {
            entries.ManagedRules = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `ManagedRules.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_PutMetricAlarmInput = (input, context) => {
    const entries = {};
    if (input.AlarmName != null) {
        entries["AlarmName"] = input.AlarmName;
    }
    if (input.AlarmDescription != null) {
        entries["AlarmDescription"] = input.AlarmDescription;
    }
    if (input.ActionsEnabled != null) {
        entries["ActionsEnabled"] = input.ActionsEnabled;
    }
    if (input.OKActions != null) {
        const memberEntries = se_ResourceList(input.OKActions, context);
        if (input.OKActions?.length === 0) {
            entries.OKActions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `OKActions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.AlarmActions != null) {
        const memberEntries = se_ResourceList(input.AlarmActions, context);
        if (input.AlarmActions?.length === 0) {
            entries.AlarmActions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `AlarmActions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.InsufficientDataActions != null) {
        const memberEntries = se_ResourceList(input.InsufficientDataActions, context);
        if (input.InsufficientDataActions?.length === 0) {
            entries.InsufficientDataActions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `InsufficientDataActions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.Statistic != null) {
        entries["Statistic"] = input.Statistic;
    }
    if (input.ExtendedStatistic != null) {
        entries["ExtendedStatistic"] = input.ExtendedStatistic;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Period != null) {
        entries["Period"] = input.Period;
    }
    if (input.Unit != null) {
        entries["Unit"] = input.Unit;
    }
    if (input.EvaluationPeriods != null) {
        entries["EvaluationPeriods"] = input.EvaluationPeriods;
    }
    if (input.DatapointsToAlarm != null) {
        entries["DatapointsToAlarm"] = input.DatapointsToAlarm;
    }
    if (input.Threshold != null) {
        entries["Threshold"] = (0, smithy_client_1.serializeFloat)(input.Threshold);
    }
    if (input.ComparisonOperator != null) {
        entries["ComparisonOperator"] = input.ComparisonOperator;
    }
    if (input.TreatMissingData != null) {
        entries["TreatMissingData"] = input.TreatMissingData;
    }
    if (input.EvaluateLowSampleCountPercentile != null) {
        entries["EvaluateLowSampleCountPercentile"] = input.EvaluateLowSampleCountPercentile;
    }
    if (input.Metrics != null) {
        const memberEntries = se_MetricDataQueries(input.Metrics, context);
        if (input.Metrics?.length === 0) {
            entries.Metrics = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Metrics.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Tags != null) {
        const memberEntries = se_TagList(input.Tags, context);
        if (input.Tags?.length === 0) {
            entries.Tags = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Tags.${key}`;
            entries[loc] = value;
        });
    }
    if (input.ThresholdMetricId != null) {
        entries["ThresholdMetricId"] = input.ThresholdMetricId;
    }
    return entries;
};
const se_PutMetricDataInput = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricData != null) {
        const memberEntries = se_MetricData(input.MetricData, context);
        if (input.MetricData?.length === 0) {
            entries.MetricData = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `MetricData.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_PutMetricStreamInput = (input, context) => {
    const entries = {};
    if (input.Name != null) {
        entries["Name"] = input.Name;
    }
    if (input.IncludeFilters != null) {
        const memberEntries = se_MetricStreamFilters(input.IncludeFilters, context);
        if (input.IncludeFilters?.length === 0) {
            entries.IncludeFilters = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `IncludeFilters.${key}`;
            entries[loc] = value;
        });
    }
    if (input.ExcludeFilters != null) {
        const memberEntries = se_MetricStreamFilters(input.ExcludeFilters, context);
        if (input.ExcludeFilters?.length === 0) {
            entries.ExcludeFilters = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `ExcludeFilters.${key}`;
            entries[loc] = value;
        });
    }
    if (input.FirehoseArn != null) {
        entries["FirehoseArn"] = input.FirehoseArn;
    }
    if (input.RoleArn != null) {
        entries["RoleArn"] = input.RoleArn;
    }
    if (input.OutputFormat != null) {
        entries["OutputFormat"] = input.OutputFormat;
    }
    if (input.Tags != null) {
        const memberEntries = se_TagList(input.Tags, context);
        if (input.Tags?.length === 0) {
            entries.Tags = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Tags.${key}`;
            entries[loc] = value;
        });
    }
    if (input.StatisticsConfigurations != null) {
        const memberEntries = se_MetricStreamStatisticsConfigurations(input.StatisticsConfigurations, context);
        if (input.StatisticsConfigurations?.length === 0) {
            entries.StatisticsConfigurations = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `StatisticsConfigurations.${key}`;
            entries[loc] = value;
        });
    }
    if (input.IncludeLinkedAccountsMetrics != null) {
        entries["IncludeLinkedAccountsMetrics"] = input.IncludeLinkedAccountsMetrics;
    }
    return entries;
};
const se_Range = (input, context) => {
    const entries = {};
    if (input.StartTime != null) {
        entries["StartTime"] = input.StartTime.toISOString().split(".")[0] + "Z";
    }
    if (input.EndTime != null) {
        entries["EndTime"] = input.EndTime.toISOString().split(".")[0] + "Z";
    }
    return entries;
};
const se_ResourceList = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_SetAlarmStateInput = (input, context) => {
    const entries = {};
    if (input.AlarmName != null) {
        entries["AlarmName"] = input.AlarmName;
    }
    if (input.StateValue != null) {
        entries["StateValue"] = input.StateValue;
    }
    if (input.StateReason != null) {
        entries["StateReason"] = input.StateReason;
    }
    if (input.StateReasonData != null) {
        entries["StateReasonData"] = input.StateReasonData;
    }
    return entries;
};
const se_SingleMetricAnomalyDetector = (input, context) => {
    const entries = {};
    if (input.Namespace != null) {
        entries["Namespace"] = input.Namespace;
    }
    if (input.MetricName != null) {
        entries["MetricName"] = input.MetricName;
    }
    if (input.Dimensions != null) {
        const memberEntries = se_Dimensions(input.Dimensions, context);
        if (input.Dimensions?.length === 0) {
            entries.Dimensions = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Dimensions.${key}`;
            entries[loc] = value;
        });
    }
    if (input.Stat != null) {
        entries["Stat"] = input.Stat;
    }
    return entries;
};
const se_StartMetricStreamsInput = (input, context) => {
    const entries = {};
    if (input.Names != null) {
        const memberEntries = se_MetricStreamNames(input.Names, context);
        if (input.Names?.length === 0) {
            entries.Names = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Names.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_Statistics = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_StatisticSet = (input, context) => {
    const entries = {};
    if (input.SampleCount != null) {
        entries["SampleCount"] = (0, smithy_client_1.serializeFloat)(input.SampleCount);
    }
    if (input.Sum != null) {
        entries["Sum"] = (0, smithy_client_1.serializeFloat)(input.Sum);
    }
    if (input.Minimum != null) {
        entries["Minimum"] = (0, smithy_client_1.serializeFloat)(input.Minimum);
    }
    if (input.Maximum != null) {
        entries["Maximum"] = (0, smithy_client_1.serializeFloat)(input.Maximum);
    }
    return entries;
};
const se_StopMetricStreamsInput = (input, context) => {
    const entries = {};
    if (input.Names != null) {
        const memberEntries = se_MetricStreamNames(input.Names, context);
        if (input.Names?.length === 0) {
            entries.Names = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Names.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_Tag = (input, context) => {
    const entries = {};
    if (input.Key != null) {
        entries["Key"] = input.Key;
    }
    if (input.Value != null) {
        entries["Value"] = input.Value;
    }
    return entries;
};
const se_TagKeyList = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = entry;
        counter++;
    }
    return entries;
};
const se_TagList = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        const memberEntries = se_Tag(entry, context);
        Object.entries(memberEntries).forEach(([key, value]) => {
            entries[`member.${counter}.${key}`] = value;
        });
        counter++;
    }
    return entries;
};
const se_TagResourceInput = (input, context) => {
    const entries = {};
    if (input.ResourceARN != null) {
        entries["ResourceARN"] = input.ResourceARN;
    }
    if (input.Tags != null) {
        const memberEntries = se_TagList(input.Tags, context);
        if (input.Tags?.length === 0) {
            entries.Tags = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `Tags.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_UntagResourceInput = (input, context) => {
    const entries = {};
    if (input.ResourceARN != null) {
        entries["ResourceARN"] = input.ResourceARN;
    }
    if (input.TagKeys != null) {
        const memberEntries = se_TagKeyList(input.TagKeys, context);
        if (input.TagKeys?.length === 0) {
            entries.TagKeys = [];
        }
        Object.entries(memberEntries).forEach(([key, value]) => {
            const loc = `TagKeys.${key}`;
            entries[loc] = value;
        });
    }
    return entries;
};
const se_Values = (input, context) => {
    const entries = {};
    let counter = 1;
    for (const entry of input) {
        if (entry === null) {
            continue;
        }
        entries[`member.${counter}`] = (0, smithy_client_1.serializeFloat)(entry);
        counter++;
    }
    return entries;
};
const de_AlarmHistoryItem = (output, context) => {
    const contents = {};
    if (output["AlarmName"] !== undefined) {
        contents.AlarmName = (0, smithy_client_1.expectString)(output["AlarmName"]);
    }
    if (output["AlarmType"] !== undefined) {
        contents.AlarmType = (0, smithy_client_1.expectString)(output["AlarmType"]);
    }
    if (output["Timestamp"] !== undefined) {
        contents.Timestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["Timestamp"]));
    }
    if (output["HistoryItemType"] !== undefined) {
        contents.HistoryItemType = (0, smithy_client_1.expectString)(output["HistoryItemType"]);
    }
    if (output["HistorySummary"] !== undefined) {
        contents.HistorySummary = (0, smithy_client_1.expectString)(output["HistorySummary"]);
    }
    if (output["HistoryData"] !== undefined) {
        contents.HistoryData = (0, smithy_client_1.expectString)(output["HistoryData"]);
    }
    return contents;
};
const de_AlarmHistoryItems = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_AlarmHistoryItem(entry, context);
    });
};
const de_AnomalyDetector = (output, context) => {
    const contents = {};
    if (output["Namespace"] !== undefined) {
        contents.Namespace = (0, smithy_client_1.expectString)(output["Namespace"]);
    }
    if (output["MetricName"] !== undefined) {
        contents.MetricName = (0, smithy_client_1.expectString)(output["MetricName"]);
    }
    if (output.Dimensions === "") {
        contents.Dimensions = [];
    }
    else if (output["Dimensions"] !== undefined && output["Dimensions"]["member"] !== undefined) {
        contents.Dimensions = de_Dimensions((0, smithy_client_1.getArrayIfSingleItem)(output["Dimensions"]["member"]), context);
    }
    if (output["Stat"] !== undefined) {
        contents.Stat = (0, smithy_client_1.expectString)(output["Stat"]);
    }
    if (output["Configuration"] !== undefined) {
        contents.Configuration = de_AnomalyDetectorConfiguration(output["Configuration"], context);
    }
    if (output["StateValue"] !== undefined) {
        contents.StateValue = (0, smithy_client_1.expectString)(output["StateValue"]);
    }
    if (output["SingleMetricAnomalyDetector"] !== undefined) {
        contents.SingleMetricAnomalyDetector = de_SingleMetricAnomalyDetector(output["SingleMetricAnomalyDetector"], context);
    }
    if (output["MetricMathAnomalyDetector"] !== undefined) {
        contents.MetricMathAnomalyDetector = de_MetricMathAnomalyDetector(output["MetricMathAnomalyDetector"], context);
    }
    return contents;
};
const de_AnomalyDetectorConfiguration = (output, context) => {
    const contents = {};
    if (output.ExcludedTimeRanges === "") {
        contents.ExcludedTimeRanges = [];
    }
    else if (output["ExcludedTimeRanges"] !== undefined && output["ExcludedTimeRanges"]["member"] !== undefined) {
        contents.ExcludedTimeRanges = de_AnomalyDetectorExcludedTimeRanges((0, smithy_client_1.getArrayIfSingleItem)(output["ExcludedTimeRanges"]["member"]), context);
    }
    if (output["MetricTimezone"] !== undefined) {
        contents.MetricTimezone = (0, smithy_client_1.expectString)(output["MetricTimezone"]);
    }
    return contents;
};
const de_AnomalyDetectorExcludedTimeRanges = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Range(entry, context);
    });
};
const de_AnomalyDetectors = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_AnomalyDetector(entry, context);
    });
};
const de_BatchFailures = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_PartialFailure(entry, context);
    });
};
const de_CompositeAlarm = (output, context) => {
    const contents = {};
    if (output["ActionsEnabled"] !== undefined) {
        contents.ActionsEnabled = (0, smithy_client_1.parseBoolean)(output["ActionsEnabled"]);
    }
    if (output.AlarmActions === "") {
        contents.AlarmActions = [];
    }
    else if (output["AlarmActions"] !== undefined && output["AlarmActions"]["member"] !== undefined) {
        contents.AlarmActions = de_ResourceList((0, smithy_client_1.getArrayIfSingleItem)(output["AlarmActions"]["member"]), context);
    }
    if (output["AlarmArn"] !== undefined) {
        contents.AlarmArn = (0, smithy_client_1.expectString)(output["AlarmArn"]);
    }
    if (output["AlarmConfigurationUpdatedTimestamp"] !== undefined) {
        contents.AlarmConfigurationUpdatedTimestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["AlarmConfigurationUpdatedTimestamp"]));
    }
    if (output["AlarmDescription"] !== undefined) {
        contents.AlarmDescription = (0, smithy_client_1.expectString)(output["AlarmDescription"]);
    }
    if (output["AlarmName"] !== undefined) {
        contents.AlarmName = (0, smithy_client_1.expectString)(output["AlarmName"]);
    }
    if (output["AlarmRule"] !== undefined) {
        contents.AlarmRule = (0, smithy_client_1.expectString)(output["AlarmRule"]);
    }
    if (output.InsufficientDataActions === "") {
        contents.InsufficientDataActions = [];
    }
    else if (output["InsufficientDataActions"] !== undefined &&
        output["InsufficientDataActions"]["member"] !== undefined) {
        contents.InsufficientDataActions = de_ResourceList((0, smithy_client_1.getArrayIfSingleItem)(output["InsufficientDataActions"]["member"]), context);
    }
    if (output.OKActions === "") {
        contents.OKActions = [];
    }
    else if (output["OKActions"] !== undefined && output["OKActions"]["member"] !== undefined) {
        contents.OKActions = de_ResourceList((0, smithy_client_1.getArrayIfSingleItem)(output["OKActions"]["member"]), context);
    }
    if (output["StateReason"] !== undefined) {
        contents.StateReason = (0, smithy_client_1.expectString)(output["StateReason"]);
    }
    if (output["StateReasonData"] !== undefined) {
        contents.StateReasonData = (0, smithy_client_1.expectString)(output["StateReasonData"]);
    }
    if (output["StateUpdatedTimestamp"] !== undefined) {
        contents.StateUpdatedTimestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["StateUpdatedTimestamp"]));
    }
    if (output["StateValue"] !== undefined) {
        contents.StateValue = (0, smithy_client_1.expectString)(output["StateValue"]);
    }
    if (output["StateTransitionedTimestamp"] !== undefined) {
        contents.StateTransitionedTimestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["StateTransitionedTimestamp"]));
    }
    if (output["ActionsSuppressedBy"] !== undefined) {
        contents.ActionsSuppressedBy = (0, smithy_client_1.expectString)(output["ActionsSuppressedBy"]);
    }
    if (output["ActionsSuppressedReason"] !== undefined) {
        contents.ActionsSuppressedReason = (0, smithy_client_1.expectString)(output["ActionsSuppressedReason"]);
    }
    if (output["ActionsSuppressor"] !== undefined) {
        contents.ActionsSuppressor = (0, smithy_client_1.expectString)(output["ActionsSuppressor"]);
    }
    if (output["ActionsSuppressorWaitPeriod"] !== undefined) {
        contents.ActionsSuppressorWaitPeriod = (0, smithy_client_1.strictParseInt32)(output["ActionsSuppressorWaitPeriod"]);
    }
    if (output["ActionsSuppressorExtensionPeriod"] !== undefined) {
        contents.ActionsSuppressorExtensionPeriod = (0, smithy_client_1.strictParseInt32)(output["ActionsSuppressorExtensionPeriod"]);
    }
    return contents;
};
const de_CompositeAlarms = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_CompositeAlarm(entry, context);
    });
};
const de_ConcurrentModificationException = (output, context) => {
    const contents = {};
    if (output["Message"] !== undefined) {
        contents.Message = (0, smithy_client_1.expectString)(output["Message"]);
    }
    return contents;
};
const de_DashboardEntries = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_DashboardEntry(entry, context);
    });
};
const de_DashboardEntry = (output, context) => {
    const contents = {};
    if (output["DashboardName"] !== undefined) {
        contents.DashboardName = (0, smithy_client_1.expectString)(output["DashboardName"]);
    }
    if (output["DashboardArn"] !== undefined) {
        contents.DashboardArn = (0, smithy_client_1.expectString)(output["DashboardArn"]);
    }
    if (output["LastModified"] !== undefined) {
        contents.LastModified = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["LastModified"]));
    }
    if (output["Size"] !== undefined) {
        contents.Size = (0, smithy_client_1.strictParseLong)(output["Size"]);
    }
    return contents;
};
const de_DashboardInvalidInputError = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    if (output.dashboardValidationMessages === "") {
        contents.dashboardValidationMessages = [];
    }
    else if (output["dashboardValidationMessages"] !== undefined &&
        output["dashboardValidationMessages"]["member"] !== undefined) {
        contents.dashboardValidationMessages = de_DashboardValidationMessages((0, smithy_client_1.getArrayIfSingleItem)(output["dashboardValidationMessages"]["member"]), context);
    }
    return contents;
};
const de_DashboardNotFoundError = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_DashboardValidationMessage = (output, context) => {
    const contents = {};
    if (output["DataPath"] !== undefined) {
        contents.DataPath = (0, smithy_client_1.expectString)(output["DataPath"]);
    }
    if (output["Message"] !== undefined) {
        contents.Message = (0, smithy_client_1.expectString)(output["Message"]);
    }
    return contents;
};
const de_DashboardValidationMessages = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_DashboardValidationMessage(entry, context);
    });
};
const de_Datapoint = (output, context) => {
    const contents = {};
    if (output["Timestamp"] !== undefined) {
        contents.Timestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["Timestamp"]));
    }
    if (output["SampleCount"] !== undefined) {
        contents.SampleCount = (0, smithy_client_1.strictParseFloat)(output["SampleCount"]);
    }
    if (output["Average"] !== undefined) {
        contents.Average = (0, smithy_client_1.strictParseFloat)(output["Average"]);
    }
    if (output["Sum"] !== undefined) {
        contents.Sum = (0, smithy_client_1.strictParseFloat)(output["Sum"]);
    }
    if (output["Minimum"] !== undefined) {
        contents.Minimum = (0, smithy_client_1.strictParseFloat)(output["Minimum"]);
    }
    if (output["Maximum"] !== undefined) {
        contents.Maximum = (0, smithy_client_1.strictParseFloat)(output["Maximum"]);
    }
    if (output["Unit"] !== undefined) {
        contents.Unit = (0, smithy_client_1.expectString)(output["Unit"]);
    }
    if (output.ExtendedStatistics === "") {
        contents.ExtendedStatistics = {};
    }
    else if (output["ExtendedStatistics"] !== undefined && output["ExtendedStatistics"]["entry"] !== undefined) {
        contents.ExtendedStatistics = de_DatapointValueMap((0, smithy_client_1.getArrayIfSingleItem)(output["ExtendedStatistics"]["entry"]), context);
    }
    return contents;
};
const de_Datapoints = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Datapoint(entry, context);
    });
};
const de_DatapointValueMap = (output, context) => {
    return output.reduce((acc, pair) => {
        if (pair["value"] === null) {
            return acc;
        }
        acc[pair["key"]] = (0, smithy_client_1.strictParseFloat)(pair["value"]);
        return acc;
    }, {});
};
const de_DatapointValues = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.strictParseFloat)(entry);
    });
};
const de_DeleteAnomalyDetectorOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_DeleteDashboardsOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_DeleteInsightRulesOutput = (output, context) => {
    const contents = {};
    if (output.Failures === "") {
        contents.Failures = [];
    }
    else if (output["Failures"] !== undefined && output["Failures"]["member"] !== undefined) {
        contents.Failures = de_BatchFailures((0, smithy_client_1.getArrayIfSingleItem)(output["Failures"]["member"]), context);
    }
    return contents;
};
const de_DeleteMetricStreamOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_DescribeAlarmHistoryOutput = (output, context) => {
    const contents = {};
    if (output.AlarmHistoryItems === "") {
        contents.AlarmHistoryItems = [];
    }
    else if (output["AlarmHistoryItems"] !== undefined && output["AlarmHistoryItems"]["member"] !== undefined) {
        contents.AlarmHistoryItems = de_AlarmHistoryItems((0, smithy_client_1.getArrayIfSingleItem)(output["AlarmHistoryItems"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    return contents;
};
const de_DescribeAlarmsForMetricOutput = (output, context) => {
    const contents = {};
    if (output.MetricAlarms === "") {
        contents.MetricAlarms = [];
    }
    else if (output["MetricAlarms"] !== undefined && output["MetricAlarms"]["member"] !== undefined) {
        contents.MetricAlarms = de_MetricAlarms((0, smithy_client_1.getArrayIfSingleItem)(output["MetricAlarms"]["member"]), context);
    }
    return contents;
};
const de_DescribeAlarmsOutput = (output, context) => {
    const contents = {};
    if (output.CompositeAlarms === "") {
        contents.CompositeAlarms = [];
    }
    else if (output["CompositeAlarms"] !== undefined && output["CompositeAlarms"]["member"] !== undefined) {
        contents.CompositeAlarms = de_CompositeAlarms((0, smithy_client_1.getArrayIfSingleItem)(output["CompositeAlarms"]["member"]), context);
    }
    if (output.MetricAlarms === "") {
        contents.MetricAlarms = [];
    }
    else if (output["MetricAlarms"] !== undefined && output["MetricAlarms"]["member"] !== undefined) {
        contents.MetricAlarms = de_MetricAlarms((0, smithy_client_1.getArrayIfSingleItem)(output["MetricAlarms"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    return contents;
};
const de_DescribeAnomalyDetectorsOutput = (output, context) => {
    const contents = {};
    if (output.AnomalyDetectors === "") {
        contents.AnomalyDetectors = [];
    }
    else if (output["AnomalyDetectors"] !== undefined && output["AnomalyDetectors"]["member"] !== undefined) {
        contents.AnomalyDetectors = de_AnomalyDetectors((0, smithy_client_1.getArrayIfSingleItem)(output["AnomalyDetectors"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    return contents;
};
const de_DescribeInsightRulesOutput = (output, context) => {
    const contents = {};
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    if (output.InsightRules === "") {
        contents.InsightRules = [];
    }
    else if (output["InsightRules"] !== undefined && output["InsightRules"]["member"] !== undefined) {
        contents.InsightRules = de_InsightRules((0, smithy_client_1.getArrayIfSingleItem)(output["InsightRules"]["member"]), context);
    }
    return contents;
};
const de_Dimension = (output, context) => {
    const contents = {};
    if (output["Name"] !== undefined) {
        contents.Name = (0, smithy_client_1.expectString)(output["Name"]);
    }
    if (output["Value"] !== undefined) {
        contents.Value = (0, smithy_client_1.expectString)(output["Value"]);
    }
    return contents;
};
const de_Dimensions = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Dimension(entry, context);
    });
};
const de_DisableInsightRulesOutput = (output, context) => {
    const contents = {};
    if (output.Failures === "") {
        contents.Failures = [];
    }
    else if (output["Failures"] !== undefined && output["Failures"]["member"] !== undefined) {
        contents.Failures = de_BatchFailures((0, smithy_client_1.getArrayIfSingleItem)(output["Failures"]["member"]), context);
    }
    return contents;
};
const de_EnableInsightRulesOutput = (output, context) => {
    const contents = {};
    if (output.Failures === "") {
        contents.Failures = [];
    }
    else if (output["Failures"] !== undefined && output["Failures"]["member"] !== undefined) {
        contents.Failures = de_BatchFailures((0, smithy_client_1.getArrayIfSingleItem)(output["Failures"]["member"]), context);
    }
    return contents;
};
const de_GetDashboardOutput = (output, context) => {
    const contents = {};
    if (output["DashboardArn"] !== undefined) {
        contents.DashboardArn = (0, smithy_client_1.expectString)(output["DashboardArn"]);
    }
    if (output["DashboardBody"] !== undefined) {
        contents.DashboardBody = (0, smithy_client_1.expectString)(output["DashboardBody"]);
    }
    if (output["DashboardName"] !== undefined) {
        contents.DashboardName = (0, smithy_client_1.expectString)(output["DashboardName"]);
    }
    return contents;
};
const de_GetInsightRuleReportOutput = (output, context) => {
    const contents = {};
    if (output.KeyLabels === "") {
        contents.KeyLabels = [];
    }
    else if (output["KeyLabels"] !== undefined && output["KeyLabels"]["member"] !== undefined) {
        contents.KeyLabels = de_InsightRuleContributorKeyLabels((0, smithy_client_1.getArrayIfSingleItem)(output["KeyLabels"]["member"]), context);
    }
    if (output["AggregationStatistic"] !== undefined) {
        contents.AggregationStatistic = (0, smithy_client_1.expectString)(output["AggregationStatistic"]);
    }
    if (output["AggregateValue"] !== undefined) {
        contents.AggregateValue = (0, smithy_client_1.strictParseFloat)(output["AggregateValue"]);
    }
    if (output["ApproximateUniqueCount"] !== undefined) {
        contents.ApproximateUniqueCount = (0, smithy_client_1.strictParseLong)(output["ApproximateUniqueCount"]);
    }
    if (output.Contributors === "") {
        contents.Contributors = [];
    }
    else if (output["Contributors"] !== undefined && output["Contributors"]["member"] !== undefined) {
        contents.Contributors = de_InsightRuleContributors((0, smithy_client_1.getArrayIfSingleItem)(output["Contributors"]["member"]), context);
    }
    if (output.MetricDatapoints === "") {
        contents.MetricDatapoints = [];
    }
    else if (output["MetricDatapoints"] !== undefined && output["MetricDatapoints"]["member"] !== undefined) {
        contents.MetricDatapoints = de_InsightRuleMetricDatapoints((0, smithy_client_1.getArrayIfSingleItem)(output["MetricDatapoints"]["member"]), context);
    }
    return contents;
};
const de_GetMetricDataOutput = (output, context) => {
    const contents = {};
    if (output.MetricDataResults === "") {
        contents.MetricDataResults = [];
    }
    else if (output["MetricDataResults"] !== undefined && output["MetricDataResults"]["member"] !== undefined) {
        contents.MetricDataResults = de_MetricDataResults((0, smithy_client_1.getArrayIfSingleItem)(output["MetricDataResults"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    if (output.Messages === "") {
        contents.Messages = [];
    }
    else if (output["Messages"] !== undefined && output["Messages"]["member"] !== undefined) {
        contents.Messages = de_MetricDataResultMessages((0, smithy_client_1.getArrayIfSingleItem)(output["Messages"]["member"]), context);
    }
    return contents;
};
const de_GetMetricStatisticsOutput = (output, context) => {
    const contents = {};
    if (output["Label"] !== undefined) {
        contents.Label = (0, smithy_client_1.expectString)(output["Label"]);
    }
    if (output.Datapoints === "") {
        contents.Datapoints = [];
    }
    else if (output["Datapoints"] !== undefined && output["Datapoints"]["member"] !== undefined) {
        contents.Datapoints = de_Datapoints((0, smithy_client_1.getArrayIfSingleItem)(output["Datapoints"]["member"]), context);
    }
    return contents;
};
const de_GetMetricStreamOutput = (output, context) => {
    const contents = {};
    if (output["Arn"] !== undefined) {
        contents.Arn = (0, smithy_client_1.expectString)(output["Arn"]);
    }
    if (output["Name"] !== undefined) {
        contents.Name = (0, smithy_client_1.expectString)(output["Name"]);
    }
    if (output.IncludeFilters === "") {
        contents.IncludeFilters = [];
    }
    else if (output["IncludeFilters"] !== undefined && output["IncludeFilters"]["member"] !== undefined) {
        contents.IncludeFilters = de_MetricStreamFilters((0, smithy_client_1.getArrayIfSingleItem)(output["IncludeFilters"]["member"]), context);
    }
    if (output.ExcludeFilters === "") {
        contents.ExcludeFilters = [];
    }
    else if (output["ExcludeFilters"] !== undefined && output["ExcludeFilters"]["member"] !== undefined) {
        contents.ExcludeFilters = de_MetricStreamFilters((0, smithy_client_1.getArrayIfSingleItem)(output["ExcludeFilters"]["member"]), context);
    }
    if (output["FirehoseArn"] !== undefined) {
        contents.FirehoseArn = (0, smithy_client_1.expectString)(output["FirehoseArn"]);
    }
    if (output["RoleArn"] !== undefined) {
        contents.RoleArn = (0, smithy_client_1.expectString)(output["RoleArn"]);
    }
    if (output["State"] !== undefined) {
        contents.State = (0, smithy_client_1.expectString)(output["State"]);
    }
    if (output["CreationDate"] !== undefined) {
        contents.CreationDate = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["CreationDate"]));
    }
    if (output["LastUpdateDate"] !== undefined) {
        contents.LastUpdateDate = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["LastUpdateDate"]));
    }
    if (output["OutputFormat"] !== undefined) {
        contents.OutputFormat = (0, smithy_client_1.expectString)(output["OutputFormat"]);
    }
    if (output.StatisticsConfigurations === "") {
        contents.StatisticsConfigurations = [];
    }
    else if (output["StatisticsConfigurations"] !== undefined &&
        output["StatisticsConfigurations"]["member"] !== undefined) {
        contents.StatisticsConfigurations = de_MetricStreamStatisticsConfigurations((0, smithy_client_1.getArrayIfSingleItem)(output["StatisticsConfigurations"]["member"]), context);
    }
    if (output["IncludeLinkedAccountsMetrics"] !== undefined) {
        contents.IncludeLinkedAccountsMetrics = (0, smithy_client_1.parseBoolean)(output["IncludeLinkedAccountsMetrics"]);
    }
    return contents;
};
const de_GetMetricWidgetImageOutput = (output, context) => {
    const contents = {};
    if (output["MetricWidgetImage"] !== undefined) {
        contents.MetricWidgetImage = context.base64Decoder(output["MetricWidgetImage"]);
    }
    return contents;
};
const de_InsightRule = (output, context) => {
    const contents = {};
    if (output["Name"] !== undefined) {
        contents.Name = (0, smithy_client_1.expectString)(output["Name"]);
    }
    if (output["State"] !== undefined) {
        contents.State = (0, smithy_client_1.expectString)(output["State"]);
    }
    if (output["Schema"] !== undefined) {
        contents.Schema = (0, smithy_client_1.expectString)(output["Schema"]);
    }
    if (output["Definition"] !== undefined) {
        contents.Definition = (0, smithy_client_1.expectString)(output["Definition"]);
    }
    if (output["ManagedRule"] !== undefined) {
        contents.ManagedRule = (0, smithy_client_1.parseBoolean)(output["ManagedRule"]);
    }
    return contents;
};
const de_InsightRuleContributor = (output, context) => {
    const contents = {};
    if (output.Keys === "") {
        contents.Keys = [];
    }
    else if (output["Keys"] !== undefined && output["Keys"]["member"] !== undefined) {
        contents.Keys = de_InsightRuleContributorKeys((0, smithy_client_1.getArrayIfSingleItem)(output["Keys"]["member"]), context);
    }
    if (output["ApproximateAggregateValue"] !== undefined) {
        contents.ApproximateAggregateValue = (0, smithy_client_1.strictParseFloat)(output["ApproximateAggregateValue"]);
    }
    if (output.Datapoints === "") {
        contents.Datapoints = [];
    }
    else if (output["Datapoints"] !== undefined && output["Datapoints"]["member"] !== undefined) {
        contents.Datapoints = de_InsightRuleContributorDatapoints((0, smithy_client_1.getArrayIfSingleItem)(output["Datapoints"]["member"]), context);
    }
    return contents;
};
const de_InsightRuleContributorDatapoint = (output, context) => {
    const contents = {};
    if (output["Timestamp"] !== undefined) {
        contents.Timestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["Timestamp"]));
    }
    if (output["ApproximateValue"] !== undefined) {
        contents.ApproximateValue = (0, smithy_client_1.strictParseFloat)(output["ApproximateValue"]);
    }
    return contents;
};
const de_InsightRuleContributorDatapoints = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_InsightRuleContributorDatapoint(entry, context);
    });
};
const de_InsightRuleContributorKeyLabels = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectString)(entry);
    });
};
const de_InsightRuleContributorKeys = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectString)(entry);
    });
};
const de_InsightRuleContributors = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_InsightRuleContributor(entry, context);
    });
};
const de_InsightRuleMetricDatapoint = (output, context) => {
    const contents = {};
    if (output["Timestamp"] !== undefined) {
        contents.Timestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["Timestamp"]));
    }
    if (output["UniqueContributors"] !== undefined) {
        contents.UniqueContributors = (0, smithy_client_1.strictParseFloat)(output["UniqueContributors"]);
    }
    if (output["MaxContributorValue"] !== undefined) {
        contents.MaxContributorValue = (0, smithy_client_1.strictParseFloat)(output["MaxContributorValue"]);
    }
    if (output["SampleCount"] !== undefined) {
        contents.SampleCount = (0, smithy_client_1.strictParseFloat)(output["SampleCount"]);
    }
    if (output["Average"] !== undefined) {
        contents.Average = (0, smithy_client_1.strictParseFloat)(output["Average"]);
    }
    if (output["Sum"] !== undefined) {
        contents.Sum = (0, smithy_client_1.strictParseFloat)(output["Sum"]);
    }
    if (output["Minimum"] !== undefined) {
        contents.Minimum = (0, smithy_client_1.strictParseFloat)(output["Minimum"]);
    }
    if (output["Maximum"] !== undefined) {
        contents.Maximum = (0, smithy_client_1.strictParseFloat)(output["Maximum"]);
    }
    return contents;
};
const de_InsightRuleMetricDatapoints = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_InsightRuleMetricDatapoint(entry, context);
    });
};
const de_InsightRules = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_InsightRule(entry, context);
    });
};
const de_InternalServiceFault = (output, context) => {
    const contents = {};
    if (output["Message"] !== undefined) {
        contents.Message = (0, smithy_client_1.expectString)(output["Message"]);
    }
    return contents;
};
const de_InvalidFormatFault = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_InvalidNextToken = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_InvalidParameterCombinationException = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_InvalidParameterValueException = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_LimitExceededException = (output, context) => {
    const contents = {};
    if (output["Message"] !== undefined) {
        contents.Message = (0, smithy_client_1.expectString)(output["Message"]);
    }
    return contents;
};
const de_LimitExceededFault = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_ListDashboardsOutput = (output, context) => {
    const contents = {};
    if (output.DashboardEntries === "") {
        contents.DashboardEntries = [];
    }
    else if (output["DashboardEntries"] !== undefined && output["DashboardEntries"]["member"] !== undefined) {
        contents.DashboardEntries = de_DashboardEntries((0, smithy_client_1.getArrayIfSingleItem)(output["DashboardEntries"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    return contents;
};
const de_ListManagedInsightRulesOutput = (output, context) => {
    const contents = {};
    if (output.ManagedRules === "") {
        contents.ManagedRules = [];
    }
    else if (output["ManagedRules"] !== undefined && output["ManagedRules"]["member"] !== undefined) {
        contents.ManagedRules = de_ManagedRuleDescriptions((0, smithy_client_1.getArrayIfSingleItem)(output["ManagedRules"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    return contents;
};
const de_ListMetricsOutput = (output, context) => {
    const contents = {};
    if (output.Metrics === "") {
        contents.Metrics = [];
    }
    else if (output["Metrics"] !== undefined && output["Metrics"]["member"] !== undefined) {
        contents.Metrics = de_Metrics((0, smithy_client_1.getArrayIfSingleItem)(output["Metrics"]["member"]), context);
    }
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    if (output.OwningAccounts === "") {
        contents.OwningAccounts = [];
    }
    else if (output["OwningAccounts"] !== undefined && output["OwningAccounts"]["member"] !== undefined) {
        contents.OwningAccounts = de_OwningAccounts((0, smithy_client_1.getArrayIfSingleItem)(output["OwningAccounts"]["member"]), context);
    }
    return contents;
};
const de_ListMetricStreamsOutput = (output, context) => {
    const contents = {};
    if (output["NextToken"] !== undefined) {
        contents.NextToken = (0, smithy_client_1.expectString)(output["NextToken"]);
    }
    if (output.Entries === "") {
        contents.Entries = [];
    }
    else if (output["Entries"] !== undefined && output["Entries"]["member"] !== undefined) {
        contents.Entries = de_MetricStreamEntries((0, smithy_client_1.getArrayIfSingleItem)(output["Entries"]["member"]), context);
    }
    return contents;
};
const de_ListTagsForResourceOutput = (output, context) => {
    const contents = {};
    if (output.Tags === "") {
        contents.Tags = [];
    }
    else if (output["Tags"] !== undefined && output["Tags"]["member"] !== undefined) {
        contents.Tags = de_TagList((0, smithy_client_1.getArrayIfSingleItem)(output["Tags"]["member"]), context);
    }
    return contents;
};
const de_ManagedRuleDescription = (output, context) => {
    const contents = {};
    if (output["TemplateName"] !== undefined) {
        contents.TemplateName = (0, smithy_client_1.expectString)(output["TemplateName"]);
    }
    if (output["ResourceARN"] !== undefined) {
        contents.ResourceARN = (0, smithy_client_1.expectString)(output["ResourceARN"]);
    }
    if (output["RuleState"] !== undefined) {
        contents.RuleState = de_ManagedRuleState(output["RuleState"], context);
    }
    return contents;
};
const de_ManagedRuleDescriptions = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_ManagedRuleDescription(entry, context);
    });
};
const de_ManagedRuleState = (output, context) => {
    const contents = {};
    if (output["RuleName"] !== undefined) {
        contents.RuleName = (0, smithy_client_1.expectString)(output["RuleName"]);
    }
    if (output["State"] !== undefined) {
        contents.State = (0, smithy_client_1.expectString)(output["State"]);
    }
    return contents;
};
const de_MessageData = (output, context) => {
    const contents = {};
    if (output["Code"] !== undefined) {
        contents.Code = (0, smithy_client_1.expectString)(output["Code"]);
    }
    if (output["Value"] !== undefined) {
        contents.Value = (0, smithy_client_1.expectString)(output["Value"]);
    }
    return contents;
};
const de_Metric = (output, context) => {
    const contents = {};
    if (output["Namespace"] !== undefined) {
        contents.Namespace = (0, smithy_client_1.expectString)(output["Namespace"]);
    }
    if (output["MetricName"] !== undefined) {
        contents.MetricName = (0, smithy_client_1.expectString)(output["MetricName"]);
    }
    if (output.Dimensions === "") {
        contents.Dimensions = [];
    }
    else if (output["Dimensions"] !== undefined && output["Dimensions"]["member"] !== undefined) {
        contents.Dimensions = de_Dimensions((0, smithy_client_1.getArrayIfSingleItem)(output["Dimensions"]["member"]), context);
    }
    return contents;
};
const de_MetricAlarm = (output, context) => {
    const contents = {};
    if (output["AlarmName"] !== undefined) {
        contents.AlarmName = (0, smithy_client_1.expectString)(output["AlarmName"]);
    }
    if (output["AlarmArn"] !== undefined) {
        contents.AlarmArn = (0, smithy_client_1.expectString)(output["AlarmArn"]);
    }
    if (output["AlarmDescription"] !== undefined) {
        contents.AlarmDescription = (0, smithy_client_1.expectString)(output["AlarmDescription"]);
    }
    if (output["AlarmConfigurationUpdatedTimestamp"] !== undefined) {
        contents.AlarmConfigurationUpdatedTimestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["AlarmConfigurationUpdatedTimestamp"]));
    }
    if (output["ActionsEnabled"] !== undefined) {
        contents.ActionsEnabled = (0, smithy_client_1.parseBoolean)(output["ActionsEnabled"]);
    }
    if (output.OKActions === "") {
        contents.OKActions = [];
    }
    else if (output["OKActions"] !== undefined && output["OKActions"]["member"] !== undefined) {
        contents.OKActions = de_ResourceList((0, smithy_client_1.getArrayIfSingleItem)(output["OKActions"]["member"]), context);
    }
    if (output.AlarmActions === "") {
        contents.AlarmActions = [];
    }
    else if (output["AlarmActions"] !== undefined && output["AlarmActions"]["member"] !== undefined) {
        contents.AlarmActions = de_ResourceList((0, smithy_client_1.getArrayIfSingleItem)(output["AlarmActions"]["member"]), context);
    }
    if (output.InsufficientDataActions === "") {
        contents.InsufficientDataActions = [];
    }
    else if (output["InsufficientDataActions"] !== undefined &&
        output["InsufficientDataActions"]["member"] !== undefined) {
        contents.InsufficientDataActions = de_ResourceList((0, smithy_client_1.getArrayIfSingleItem)(output["InsufficientDataActions"]["member"]), context);
    }
    if (output["StateValue"] !== undefined) {
        contents.StateValue = (0, smithy_client_1.expectString)(output["StateValue"]);
    }
    if (output["StateReason"] !== undefined) {
        contents.StateReason = (0, smithy_client_1.expectString)(output["StateReason"]);
    }
    if (output["StateReasonData"] !== undefined) {
        contents.StateReasonData = (0, smithy_client_1.expectString)(output["StateReasonData"]);
    }
    if (output["StateUpdatedTimestamp"] !== undefined) {
        contents.StateUpdatedTimestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["StateUpdatedTimestamp"]));
    }
    if (output["MetricName"] !== undefined) {
        contents.MetricName = (0, smithy_client_1.expectString)(output["MetricName"]);
    }
    if (output["Namespace"] !== undefined) {
        contents.Namespace = (0, smithy_client_1.expectString)(output["Namespace"]);
    }
    if (output["Statistic"] !== undefined) {
        contents.Statistic = (0, smithy_client_1.expectString)(output["Statistic"]);
    }
    if (output["ExtendedStatistic"] !== undefined) {
        contents.ExtendedStatistic = (0, smithy_client_1.expectString)(output["ExtendedStatistic"]);
    }
    if (output.Dimensions === "") {
        contents.Dimensions = [];
    }
    else if (output["Dimensions"] !== undefined && output["Dimensions"]["member"] !== undefined) {
        contents.Dimensions = de_Dimensions((0, smithy_client_1.getArrayIfSingleItem)(output["Dimensions"]["member"]), context);
    }
    if (output["Period"] !== undefined) {
        contents.Period = (0, smithy_client_1.strictParseInt32)(output["Period"]);
    }
    if (output["Unit"] !== undefined) {
        contents.Unit = (0, smithy_client_1.expectString)(output["Unit"]);
    }
    if (output["EvaluationPeriods"] !== undefined) {
        contents.EvaluationPeriods = (0, smithy_client_1.strictParseInt32)(output["EvaluationPeriods"]);
    }
    if (output["DatapointsToAlarm"] !== undefined) {
        contents.DatapointsToAlarm = (0, smithy_client_1.strictParseInt32)(output["DatapointsToAlarm"]);
    }
    if (output["Threshold"] !== undefined) {
        contents.Threshold = (0, smithy_client_1.strictParseFloat)(output["Threshold"]);
    }
    if (output["ComparisonOperator"] !== undefined) {
        contents.ComparisonOperator = (0, smithy_client_1.expectString)(output["ComparisonOperator"]);
    }
    if (output["TreatMissingData"] !== undefined) {
        contents.TreatMissingData = (0, smithy_client_1.expectString)(output["TreatMissingData"]);
    }
    if (output["EvaluateLowSampleCountPercentile"] !== undefined) {
        contents.EvaluateLowSampleCountPercentile = (0, smithy_client_1.expectString)(output["EvaluateLowSampleCountPercentile"]);
    }
    if (output.Metrics === "") {
        contents.Metrics = [];
    }
    else if (output["Metrics"] !== undefined && output["Metrics"]["member"] !== undefined) {
        contents.Metrics = de_MetricDataQueries((0, smithy_client_1.getArrayIfSingleItem)(output["Metrics"]["member"]), context);
    }
    if (output["ThresholdMetricId"] !== undefined) {
        contents.ThresholdMetricId = (0, smithy_client_1.expectString)(output["ThresholdMetricId"]);
    }
    if (output["EvaluationState"] !== undefined) {
        contents.EvaluationState = (0, smithy_client_1.expectString)(output["EvaluationState"]);
    }
    if (output["StateTransitionedTimestamp"] !== undefined) {
        contents.StateTransitionedTimestamp = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["StateTransitionedTimestamp"]));
    }
    return contents;
};
const de_MetricAlarms = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricAlarm(entry, context);
    });
};
const de_MetricDataQueries = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricDataQuery(entry, context);
    });
};
const de_MetricDataQuery = (output, context) => {
    const contents = {};
    if (output["Id"] !== undefined) {
        contents.Id = (0, smithy_client_1.expectString)(output["Id"]);
    }
    if (output["MetricStat"] !== undefined) {
        contents.MetricStat = de_MetricStat(output["MetricStat"], context);
    }
    if (output["Expression"] !== undefined) {
        contents.Expression = (0, smithy_client_1.expectString)(output["Expression"]);
    }
    if (output["Label"] !== undefined) {
        contents.Label = (0, smithy_client_1.expectString)(output["Label"]);
    }
    if (output["ReturnData"] !== undefined) {
        contents.ReturnData = (0, smithy_client_1.parseBoolean)(output["ReturnData"]);
    }
    if (output["Period"] !== undefined) {
        contents.Period = (0, smithy_client_1.strictParseInt32)(output["Period"]);
    }
    if (output["AccountId"] !== undefined) {
        contents.AccountId = (0, smithy_client_1.expectString)(output["AccountId"]);
    }
    return contents;
};
const de_MetricDataResult = (output, context) => {
    const contents = {};
    if (output["Id"] !== undefined) {
        contents.Id = (0, smithy_client_1.expectString)(output["Id"]);
    }
    if (output["Label"] !== undefined) {
        contents.Label = (0, smithy_client_1.expectString)(output["Label"]);
    }
    if (output.Timestamps === "") {
        contents.Timestamps = [];
    }
    else if (output["Timestamps"] !== undefined && output["Timestamps"]["member"] !== undefined) {
        contents.Timestamps = de_Timestamps((0, smithy_client_1.getArrayIfSingleItem)(output["Timestamps"]["member"]), context);
    }
    if (output.Values === "") {
        contents.Values = [];
    }
    else if (output["Values"] !== undefined && output["Values"]["member"] !== undefined) {
        contents.Values = de_DatapointValues((0, smithy_client_1.getArrayIfSingleItem)(output["Values"]["member"]), context);
    }
    if (output["StatusCode"] !== undefined) {
        contents.StatusCode = (0, smithy_client_1.expectString)(output["StatusCode"]);
    }
    if (output.Messages === "") {
        contents.Messages = [];
    }
    else if (output["Messages"] !== undefined && output["Messages"]["member"] !== undefined) {
        contents.Messages = de_MetricDataResultMessages((0, smithy_client_1.getArrayIfSingleItem)(output["Messages"]["member"]), context);
    }
    return contents;
};
const de_MetricDataResultMessages = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MessageData(entry, context);
    });
};
const de_MetricDataResults = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricDataResult(entry, context);
    });
};
const de_MetricMathAnomalyDetector = (output, context) => {
    const contents = {};
    if (output.MetricDataQueries === "") {
        contents.MetricDataQueries = [];
    }
    else if (output["MetricDataQueries"] !== undefined && output["MetricDataQueries"]["member"] !== undefined) {
        contents.MetricDataQueries = de_MetricDataQueries((0, smithy_client_1.getArrayIfSingleItem)(output["MetricDataQueries"]["member"]), context);
    }
    return contents;
};
const de_Metrics = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Metric(entry, context);
    });
};
const de_MetricStat = (output, context) => {
    const contents = {};
    if (output["Metric"] !== undefined) {
        contents.Metric = de_Metric(output["Metric"], context);
    }
    if (output["Period"] !== undefined) {
        contents.Period = (0, smithy_client_1.strictParseInt32)(output["Period"]);
    }
    if (output["Stat"] !== undefined) {
        contents.Stat = (0, smithy_client_1.expectString)(output["Stat"]);
    }
    if (output["Unit"] !== undefined) {
        contents.Unit = (0, smithy_client_1.expectString)(output["Unit"]);
    }
    return contents;
};
const de_MetricStreamEntries = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricStreamEntry(entry, context);
    });
};
const de_MetricStreamEntry = (output, context) => {
    const contents = {};
    if (output["Arn"] !== undefined) {
        contents.Arn = (0, smithy_client_1.expectString)(output["Arn"]);
    }
    if (output["CreationDate"] !== undefined) {
        contents.CreationDate = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["CreationDate"]));
    }
    if (output["LastUpdateDate"] !== undefined) {
        contents.LastUpdateDate = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["LastUpdateDate"]));
    }
    if (output["Name"] !== undefined) {
        contents.Name = (0, smithy_client_1.expectString)(output["Name"]);
    }
    if (output["FirehoseArn"] !== undefined) {
        contents.FirehoseArn = (0, smithy_client_1.expectString)(output["FirehoseArn"]);
    }
    if (output["State"] !== undefined) {
        contents.State = (0, smithy_client_1.expectString)(output["State"]);
    }
    if (output["OutputFormat"] !== undefined) {
        contents.OutputFormat = (0, smithy_client_1.expectString)(output["OutputFormat"]);
    }
    return contents;
};
const de_MetricStreamFilter = (output, context) => {
    const contents = {};
    if (output["Namespace"] !== undefined) {
        contents.Namespace = (0, smithy_client_1.expectString)(output["Namespace"]);
    }
    if (output.MetricNames === "") {
        contents.MetricNames = [];
    }
    else if (output["MetricNames"] !== undefined && output["MetricNames"]["member"] !== undefined) {
        contents.MetricNames = de_MetricStreamFilterMetricNames((0, smithy_client_1.getArrayIfSingleItem)(output["MetricNames"]["member"]), context);
    }
    return contents;
};
const de_MetricStreamFilterMetricNames = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectString)(entry);
    });
};
const de_MetricStreamFilters = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricStreamFilter(entry, context);
    });
};
const de_MetricStreamStatisticsAdditionalStatistics = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectString)(entry);
    });
};
const de_MetricStreamStatisticsConfiguration = (output, context) => {
    const contents = {};
    if (output.IncludeMetrics === "") {
        contents.IncludeMetrics = [];
    }
    else if (output["IncludeMetrics"] !== undefined && output["IncludeMetrics"]["member"] !== undefined) {
        contents.IncludeMetrics = de_MetricStreamStatisticsIncludeMetrics((0, smithy_client_1.getArrayIfSingleItem)(output["IncludeMetrics"]["member"]), context);
    }
    if (output.AdditionalStatistics === "") {
        contents.AdditionalStatistics = [];
    }
    else if (output["AdditionalStatistics"] !== undefined && output["AdditionalStatistics"]["member"] !== undefined) {
        contents.AdditionalStatistics = de_MetricStreamStatisticsAdditionalStatistics((0, smithy_client_1.getArrayIfSingleItem)(output["AdditionalStatistics"]["member"]), context);
    }
    return contents;
};
const de_MetricStreamStatisticsConfigurations = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricStreamStatisticsConfiguration(entry, context);
    });
};
const de_MetricStreamStatisticsIncludeMetrics = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_MetricStreamStatisticsMetric(entry, context);
    });
};
const de_MetricStreamStatisticsMetric = (output, context) => {
    const contents = {};
    if (output["Namespace"] !== undefined) {
        contents.Namespace = (0, smithy_client_1.expectString)(output["Namespace"]);
    }
    if (output["MetricName"] !== undefined) {
        contents.MetricName = (0, smithy_client_1.expectString)(output["MetricName"]);
    }
    return contents;
};
const de_MissingRequiredParameterException = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_OwningAccounts = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectString)(entry);
    });
};
const de_PartialFailure = (output, context) => {
    const contents = {};
    if (output["FailureResource"] !== undefined) {
        contents.FailureResource = (0, smithy_client_1.expectString)(output["FailureResource"]);
    }
    if (output["ExceptionType"] !== undefined) {
        contents.ExceptionType = (0, smithy_client_1.expectString)(output["ExceptionType"]);
    }
    if (output["FailureCode"] !== undefined) {
        contents.FailureCode = (0, smithy_client_1.expectString)(output["FailureCode"]);
    }
    if (output["FailureDescription"] !== undefined) {
        contents.FailureDescription = (0, smithy_client_1.expectString)(output["FailureDescription"]);
    }
    return contents;
};
const de_PutAnomalyDetectorOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_PutDashboardOutput = (output, context) => {
    const contents = {};
    if (output.DashboardValidationMessages === "") {
        contents.DashboardValidationMessages = [];
    }
    else if (output["DashboardValidationMessages"] !== undefined &&
        output["DashboardValidationMessages"]["member"] !== undefined) {
        contents.DashboardValidationMessages = de_DashboardValidationMessages((0, smithy_client_1.getArrayIfSingleItem)(output["DashboardValidationMessages"]["member"]), context);
    }
    return contents;
};
const de_PutInsightRuleOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_PutManagedInsightRulesOutput = (output, context) => {
    const contents = {};
    if (output.Failures === "") {
        contents.Failures = [];
    }
    else if (output["Failures"] !== undefined && output["Failures"]["member"] !== undefined) {
        contents.Failures = de_BatchFailures((0, smithy_client_1.getArrayIfSingleItem)(output["Failures"]["member"]), context);
    }
    return contents;
};
const de_PutMetricStreamOutput = (output, context) => {
    const contents = {};
    if (output["Arn"] !== undefined) {
        contents.Arn = (0, smithy_client_1.expectString)(output["Arn"]);
    }
    return contents;
};
const de_Range = (output, context) => {
    const contents = {};
    if (output["StartTime"] !== undefined) {
        contents.StartTime = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["StartTime"]));
    }
    if (output["EndTime"] !== undefined) {
        contents.EndTime = (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(output["EndTime"]));
    }
    return contents;
};
const de_ResourceList = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectString)(entry);
    });
};
const de_ResourceNotFound = (output, context) => {
    const contents = {};
    if (output["message"] !== undefined) {
        contents.message = (0, smithy_client_1.expectString)(output["message"]);
    }
    return contents;
};
const de_ResourceNotFoundException = (output, context) => {
    const contents = {};
    if (output["ResourceType"] !== undefined) {
        contents.ResourceType = (0, smithy_client_1.expectString)(output["ResourceType"]);
    }
    if (output["ResourceId"] !== undefined) {
        contents.ResourceId = (0, smithy_client_1.expectString)(output["ResourceId"]);
    }
    if (output["Message"] !== undefined) {
        contents.Message = (0, smithy_client_1.expectString)(output["Message"]);
    }
    return contents;
};
const de_SingleMetricAnomalyDetector = (output, context) => {
    const contents = {};
    if (output["Namespace"] !== undefined) {
        contents.Namespace = (0, smithy_client_1.expectString)(output["Namespace"]);
    }
    if (output["MetricName"] !== undefined) {
        contents.MetricName = (0, smithy_client_1.expectString)(output["MetricName"]);
    }
    if (output.Dimensions === "") {
        contents.Dimensions = [];
    }
    else if (output["Dimensions"] !== undefined && output["Dimensions"]["member"] !== undefined) {
        contents.Dimensions = de_Dimensions((0, smithy_client_1.getArrayIfSingleItem)(output["Dimensions"]["member"]), context);
    }
    if (output["Stat"] !== undefined) {
        contents.Stat = (0, smithy_client_1.expectString)(output["Stat"]);
    }
    return contents;
};
const de_StartMetricStreamsOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_StopMetricStreamsOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_Tag = (output, context) => {
    const contents = {};
    if (output["Key"] !== undefined) {
        contents.Key = (0, smithy_client_1.expectString)(output["Key"]);
    }
    if (output["Value"] !== undefined) {
        contents.Value = (0, smithy_client_1.expectString)(output["Value"]);
    }
    return contents;
};
const de_TagList = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_Tag(entry, context);
    });
};
const de_TagResourceOutput = (output, context) => {
    const contents = {};
    return contents;
};
const de_Timestamps = (output, context) => {
    return (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return (0, smithy_client_1.expectNonNull)((0, smithy_client_1.parseRfc3339DateTimeWithOffset)(entry));
    });
};
const de_UntagResourceOutput = (output, context) => {
    const contents = {};
    return contents;
};
const deserializeMetadata = (output) => ({
    httpStatusCode: output.statusCode,
    requestId: output.headers["x-amzn-requestid"] ?? output.headers["x-amzn-request-id"] ?? output.headers["x-amz-request-id"],
    extendedRequestId: output.headers["x-amz-id-2"],
    cfId: output.headers["x-amz-cf-id"],
});
const collectBodyString = (streamBody, context) => (0, smithy_client_1.collectBody)(streamBody, context).then((body) => context.utf8Encoder(body));
const throwDefaultError = (0, smithy_client_1.withBaseException)(CloudWatchServiceException_1.CloudWatchServiceException);
const buildHttpRpcRequest = async (context, headers, path, resolvedHostname, body) => {
    const { hostname, protocol = "https", port, path: basePath } = await context.endpoint();
    const contents = {
        protocol,
        hostname,
        port,
        method: "POST",
        path: basePath.endsWith("/") ? basePath.slice(0, -1) + path : basePath + path,
        headers,
    };
    if (resolvedHostname !== undefined) {
        contents.hostname = resolvedHostname;
    }
    if (body !== undefined) {
        contents.body = body;
    }
    return new protocol_http_1.HttpRequest(contents);
};
const SHARED_HEADERS = {
    "content-type": "application/x-www-form-urlencoded",
};
const parseBody = (streamBody, context) => collectBodyString(streamBody, context).then((encoded) => {
    if (encoded.length) {
        const parser = new fast_xml_parser_1.XMLParser({
            attributeNamePrefix: "",
            htmlEntities: true,
            ignoreAttributes: false,
            ignoreDeclaration: true,
            parseTagValue: false,
            trimValues: false,
            tagValueProcessor: (_, val) => (val.trim() === "" && val.includes("\n") ? "" : undefined),
        });
        parser.addEntity("#xD", "\r");
        parser.addEntity("#10", "\n");
        const parsedObj = parser.parse(encoded);
        const textNodeName = "#text";
        const key = Object.keys(parsedObj)[0];
        const parsedObjToReturn = parsedObj[key];
        if (parsedObjToReturn[textNodeName]) {
            parsedObjToReturn[key] = parsedObjToReturn[textNodeName];
            delete parsedObjToReturn[textNodeName];
        }
        return (0, smithy_client_1.getValueFromTextNode)(parsedObjToReturn);
    }
    return {};
});
const parseErrorBody = async (errorBody, context) => {
    const value = await parseBody(errorBody, context);
    if (value.Error) {
        value.Error.message = value.Error.message ?? value.Error.Message;
    }
    return value;
};
const buildFormUrlencodedString = (formEntries) => Object.entries(formEntries)
    .map(([key, value]) => (0, smithy_client_1.extendedEncodeURIComponent)(key) + "=" + (0, smithy_client_1.extendedEncodeURIComponent)(value))
    .join("&");
const loadQueryErrorCode = (output, data) => {
    if (data.Error?.Code !== undefined) {
        return data.Error.Code;
    }
    if (output.statusCode == 404) {
        return "NotFound";
    }
};
